From 4aa5c9a48850597a6cd150067e9b96ed9b052748 Mon Sep 17 00:00:00 2001
From: Paul Philippov <themactep@gmail.com>
Date: Sun, 22 Sep 2024 02:13:35 -0400
Subject: [PATCH 02/10] grammar

---
 Makefile                | 2 +-
 package/pkg-download.mk | 2 +-
 2 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/Makefile b/Makefile
index 3cb10fa87d..022a713835 100644
--- a/Makefile
+++ b/Makefile
@@ -1068,7 +1068,7 @@ endif
 # Makefiles. Alternatively, if a non-empty VARS variable is passed,
 # only the variables matching the make pattern passed in VARS are
 # displayed.
-# show-vars does the same, but as a JSON dictionnary.
+# show-vars does the same, but as a JSON dictionary.
 #
 # Note: we iterate of .VARIABLES and filter each variable individually,
 # to workaround a bug in make 4.3; see https://savannah.gnu.org/bugs/?59093
diff --git a/package/pkg-download.mk b/package/pkg-download.mk
index fa1e06537e..fcb4bd2d90 100644
--- a/package/pkg-download.mk
+++ b/package/pkg-download.mk
@@ -27,7 +27,7 @@ else
 BR2_DL_DIR = $(DL_DIR)
 endif
 
-# ensure it exists and a absolute path, derefrecing symlinks
+# ensure it exists and an absolute path, dereferencing symlinks
 DL_DIR := $(shell mkdir -p $(DL_DIR) && cd $(DL_DIR) >/dev/null && pwd -P)
 
 #
-- 
2.47.2

