diff --git a/support/kconfig/confdata.c b/support/kconfig/confdata.c
index 9ab980edd8..d633395051 100644
--- a/support/kconfig/confdata.c
+++ b/support/kconfig/confdata.c
@@ -753,7 +753,7 @@ int conf_write(const char *name)
 	dirname[0] = 0;
 	if (name && name[0]) {
 		struct stat st;
-		char *slash;
+		const char *slash;
 
 		if (!stat(name, &st) && S_ISDIR(st.st_mode)) {
 			strcpy(dirname, name);
