diff --git a/os_dep/linux/custom_gpio_linux.c b/os_dep/linux/custom_gpio_linux.c
index 9658e22..3aba45d 100755
--- a/os_dep/linux/custom_gpio_linux.c
+++ b/os_dep/linux/custom_gpio_linux.c
@@ -329,6 +329,8 @@ int unregister_net_gpio_init(void)
 
 #else
 
+#if 0 /* disable gpio handling for thingino*/
+
 /************************************************/
 /*          ingenic wlan setting                */
 /************************************************/
@@ -374,5 +376,6 @@ void rtw_wifi_gpio_wlan_ctrl(int onoff)
 	return;
 }
 
+#endif
 
 #endif /* CONFIG_PLATFORM_SPRD */
diff --git a/platform/platform_ops.c b/platform/platform_ops.c
index 6d7ec42..3ec61e3 100755
--- a/platform/platform_ops.c
+++ b/platform/platform_ops.c
@@ -23,15 +23,18 @@ extern void rtw_wifi_gpio_wlan_ctrl(int onoff);
 int platform_wifi_power_on(void)
 {
 	int ret = 0;
+#if 0
     rtw_wifi_gpio_wlan_ctrl(1);
-
+#endif
 
 	return ret;
 }
 
 void platform_wifi_power_off(void)
 {
+#if 0
     rtw_wifi_gpio_wlan_ctrl(0);
+#endif
 
 }
 #endif /* !CONFIG_PLATFORM_OPS */
