From eda5f49127b2e98b2d98aa0e56d9d7a19e0faa2f Mon Sep 17 00:00:00 2001
From: gtxaspec <gtxent@gmail.com>
Date: Thu, 19 Dec 2024 22:12:07 -0800
Subject: [PATCH 07/10] uboot: update env variable for buildroot

---
 boot/uboot/uboot.mk | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/boot/uboot/uboot.mk b/boot/uboot/uboot.mk
index c78ffbe15d..628c261244 100644
--- a/boot/uboot/uboot.mk
+++ b/boot/uboot/uboot.mk
@@ -570,9 +570,9 @@ ifeq ($(BR2_TARGET_UBOOT)$(BR_BUILDING),yy)
 #
 # Patch uboot headers with env data for device if uenv.txt exists
 #
-ifneq ($(U_BOOT_ENV_TXT),)
+ifneq ($(U_BOOT_ENV_FINAL_TXT),)
 define PATCH_DEV_ENV
-	$(BR2_EXTERNAL)/scripts/uboot-device-env.sh $(U_BOOT_ENV_TXT) $(@D)/include/configs/isvp_common.h
+	$(BR2_EXTERNAL)/scripts/uboot-device-env.sh $(U_BOOT_ENV_FINAL_TXT) $(@D)/include/configs/isvp_common.h
 endef
 UBOOT_PRE_BUILD_HOOKS += PATCH_DEV_ENV
 endif
-- 
2.47.2

