From 85bc423c7416d18beb34d2564a1ef47767fb7476 Mon Sep 17 00:00:00 2001
From: Paul Philippov <themactep@gmail.com>
Date: Sun, 22 Sep 2024 02:19:38 -0400
Subject: [PATCH 09/10] package/libcurl: add path to certificates

---
 package/libcurl/libcurl.mk | 1 +
 1 file changed, 1 insertion(+)

diff --git a/package/libcurl/libcurl.mk b/package/libcurl/libcurl.mk
index 1cfdda27d2..d40279deee 100644
--- a/package/libcurl/libcurl.mk
+++ b/package/libcurl/libcurl.mk
@@ -77,6 +77,7 @@ endif
 
 ifeq ($(BR2_PACKAGE_LIBCURL_MBEDTLS),y)
 LIBCURL_CONF_OPTS += --with-mbedtls=$(STAGING_DIR)/usr
+LIBCURL_CONF_OPTS += --with-ca-bundle=/etc/ssl/certs/ca-certificates.crt
 LIBCURL_DEPENDENCIES += mbedtls
 else
 LIBCURL_CONF_OPTS += --without-mbedtls
-- 
2.47.2

