--- a/drivers/mmc/host/jzmmc_v12.c
+++ b/drivers/mmc/host/jzmmc_v12.c
@@ -957,7 +957,7 @@
 	 * It never happens in design, but we add BUG_ON here to prevent it.
 	 */
 	if ((host->state != STATE_IDLE) && (mrq->data != NULL)) {
-		dev_warn(host->dev, "operate in non-idle state = host->state\n", host->state);
+		dev_warn(host->dev, "operate in non-idle state = %s\n", host->state);
 		WARN_ON(1);
 	}
 
