diff --git a/libfaac/meson.build b/libfaac/meson.build
index 452343d..5ed1f15 100644
--- a/libfaac/meson.build
+++ b/libfaac/meson.build
@@ -42,19 +42,21 @@ libfaac = library('faac', common_src,
     install: true
 )
 
-drm_src = [
-    'kiss_fft/kiss_fftr.c',
-    'kiss_fft/kiss_fft.c',
-    'kiss_fft/kiss_fft.h'
-]
-libfaac_drm = library('faac_drm', common_src + drm_src,
-    include_directories: ['..', '../include'],
-    c_args: c_args + ['-DDRM'],
-    link_args: link_args,
-    dependencies: [ libm ],
-    vs_module_defs: 'libfaac.def',
-    install: true
-)
+if get_option('enable_drm')
+    drm_src = [
+        'kiss_fft/kiss_fftr.c',
+        'kiss_fft/kiss_fft.c',
+        'kiss_fft/kiss_fft.h'
+    ]
+    libfaac_drm = library('faac_drm', common_src + drm_src,
+        include_directories: ['..', '../include'],
+        c_args: c_args + ['-DDRM'],
+        link_args: link_args,
+        dependencies: [ libm ],
+        vs_module_defs: 'libfaac.def',
+        install: true
+    )
+endif
 
 pkgconfig = import('pkgconfig')
 pkgconfig.generate(
diff --git a/meson_options.txt b/meson_options.txt
index f52d146..d59a0f8 100644
--- a/meson_options.txt
+++ b/meson_options.txt
@@ -2,3 +2,8 @@ option('frontend',
     description: 'Build the frontend',
     type: 'boolean',
     value: true)
+
+option('enable_drm',
+    description: 'Build the DRM variant library',
+    type: 'boolean',
+    value: false)
