 libfaac/faac_real.h | 31 +++++++++++++++++++++++++++++--
 libfaac/quantize.c  |  2 +-
 2 files changed, 30 insertions(+), 3 deletions(-)

diff --git a/libfaac/faac_real.h b/libfaac/faac_real.h
index 4d004d7..a2e76fd 100644
--- a/libfaac/faac_real.h
+++ b/libfaac/faac_real.h
@@ -25,6 +25,33 @@
 #endif
 
 #include <math.h>
+#include <stdint.h>
+
+static inline float fast_log2(float x)
+{
+    union { float f; uint32_t i; } vx = { x };
+    float y = (float)vx.i;
+    y *= 1.1920928955078125e-7f;
+    return y - 126.94269504f;
+}
+
+static inline float fast_log10(float x)
+{
+    return fast_log2(x) * 0.3010299956639812f;
+}
+
+static inline float fast_pow2(float x)
+{
+    if (x < -126.0f) return 0.0f;
+    if (x > 128.0f) return 3e38f;
+    union { uint32_t i; float f; } v = { (uint32_t)((x + 126.94269504f) * 8388608.0f) };
+    return v.f;
+}
+
+static inline float fast_pow(float x, float y)
+{
+    return fast_pow2(y * fast_log2(x));
+}
 
 #ifdef FAAC_PRECISION_SINGLE
 typedef float faac_real;
@@ -32,8 +59,8 @@ typedef float faac_real;
 #define FAAC_COS cosf
 #define FAAC_SQRT sqrtf
 #define FAAC_FABS fabsf
-#define FAAC_LOG10 log10f
-#define FAAC_POW powf
+#define FAAC_LOG10  fast_log10
+#define FAAC_POW    fast_pow
 #define FAAC_ASIN asinf
 #define FAAC_LRINT lrintf
 #define FAAC_FLOOR floorf
diff --git a/libfaac/quantize.c b/libfaac/quantize.c
index 27d3a79..0f67c9e 100644
--- a/libfaac/quantize.c
+++ b/libfaac/quantize.c
@@ -171,7 +171,7 @@ static void qlevel(CoderInfo * __restrict coderInfo,
                   )
 {
     int sb;
-#if !defined(__clang__) && defined(__GNUC__) && (GCC_VERSION >= 40600)
+#if !defined(__clang__) && defined(__GNUC__) && (GCC_VERSION >= 40600) && false
     /* 2^0.25 (1.50515 dB) step from AAC specs */
     static const faac_real sfstep = 1.0 / FAAC_LOG10(FAAC_SQRT(FAAC_SQRT(2.0)));
 #else

