--- a/configure
+++ b/configure
@@ -13404,7 +13404,7 @@
      fi

 else $as_nop
-  enable_float_approx=$has_float_approx
+  enable_float_approx=yes
 fi


@@ -15639,7 +15639,7 @@
 then :
   enableval=$enable_extra_programs;
 else $as_nop
-  enable_extra_programs=yes
+  enable_extra_programs=no
 fi


--- a/configure.ac
+++ b/configure.ac
@@ -202,7 +202,7 @@
        AC_WARN([Floating point approximations are not supported on all platforms.])
      fi
     ],
-    [enable_float_approx=$has_float_approx])
+    [enable_float_approx=yes])

 AS_IF([test "$enable_float_approx" = "yes"],[
   AC_DEFINE([FLOAT_APPROX], [1], [Float approximations])
@@ -949,7 +949,7 @@

 AC_ARG_ENABLE([extra-programs],
     [AS_HELP_STRING([--disable-extra-programs], [Do not build extra programs (demo and tests)])],,
-    [enable_extra_programs=yes])
+    [enable_extra_programs=no])

 AM_CONDITIONAL([EXTRA_PROGRAMS], [test "$enable_extra_programs" = "yes"])

--- a/Makefile.am
+++ b/Makefile.am
@@ -5,6 +5,9 @@
 AUTOMAKE_OPTIONS = subdir-objects
 ACLOCAL_AMFLAGS = -I m4

+AM_CFLAGS = -ffunction-sections -fdata-sections
+AM_LDFLAGS = -Wl,--gc-sections -Wl,-z,max-page-size=0x1000
+
 lib_LTLIBRARIES = libopus.la

 DIST_SUBDIRS = doc
--- a/Makefile.in
+++ b/Makefile.in
@@ -14,7 +14,8 @@

 @SET_MAKE@

-
+AM_CFLAGS = -ffunction-sections -fdata-sections
+AM_LDFLAGS = -Wl,--gc-sections -Wl,-z,max-page-size=0x1000


 VPATH = @srcdir@
