--- a/Makefile
+++ b/Makefile
@@ -4,15 +4,15 @@
 OBJECTS_N = onvif_notify_server.o conf.o utils.o log.o ezxml_wrapper.o ezxml/ezxml.o
 OBJECTS_W = wsd_simple_server.o utils.o log.o ezxml_wrapper.o ezxml/ezxml.o
 ifdef HAVE_MBEDTLS
-INCLUDE = -DHAVE_MBEDTLS -Iextras/mbedtls/include -ffunction-sections -fdata-sections -lrt
-LIBS_O = -Wl,--gc-sections extras/mbedtls/library/libmbedcrypto.a -lpthread -lrt
-LIBS_N = -Wl,--gc-sections extras/mbedtls/library/libmbedcrypto.a -lpthread -lrt
+INCLUDE = -DHAVE_MBEDTLS -ffunction-sections -fdata-sections -lrt
+LIBS_O = $(LDFLAGS) -Wl,--gc-sections -lmbedtls -lmbedcrypto -lpthread -lrt
+LIBS_N = $(LDFLAGS) -Wl,--gc-sections -lmbedtls -lmbedcrypto -lpthread -lrt
 else
-INCLUDE = -Iextras/libtomcrypt/src/headers -ffunction-sections -fdata-sections -lrt
-LIBS_O = -Wl,--gc-sections extras/libtomcrypt/libtomcrypt.a -lpthread -lrt
-LIBS_N = -Wl,--gc-sections extras/libtomcrypt/libtomcrypt.a -lpthread -lrt
+INCLUDE = -ffunction-sections -fdata-sections -lrt
+LIBS_O = $(LDFLAGS) -Wl,--gc-sections -ltomcrypt -lpthread -lrt
+LIBS_N = $(LDFLAGS) -Wl,--gc-sections -ltomcrypt -lpthread -lrt
 endif
-LIBS_W = -Wl,--gc-sections
+LIBS_W = $(LDFLAGS) -Wl,--gc-sections
 
 ifeq ($(STRIP), )
     STRIP=echo
