--- a/Makefile
+++ b/Makefile
@@ -6,17 +6,17 @@
 OBJECTS_W = wsd_simple_server.o utils.o log.o ezxml_wrapper.o ezxml/ezxml.o
 ifdef HAVE_WOLFSSL
 INCLUDE = -DHAVE_WOLFSSL -Iextras/wolfssl -ffunction-sections -fdata-sections -lrt
-LIBS_O = -Wl,--gc-sections extras/wolfssl/src/.libs/libwolfssl.a -lpthread -lrt
-LIBS_N = -Wl,--gc-sections extras/wolfssl/src/.libs/libwolfssl.a -lpthread -lrt
+LIBS_O = -Wl,--gc-sections -lwolfssl -lpthread -lrt
+LIBS_N = -Wl,--gc-sections -lwolfssl -lpthread -lrt
 else
 ifdef HAVE_MBEDTLS
 INCLUDE = -DHAVE_MBEDTLS -Iextras/mbedtls/include -ffunction-sections -fdata-sections -lrt
-LIBS_O = -Wl,--gc-sections extras/mbedtls/library/libmbedcrypto.a -lpthread -lrt
-LIBS_N = -Wl,--gc-sections extras/mbedtls/library/libmbedcrypto.a -lpthread -lrt
+LIBS_O = -Wl,--gc-sections -lmbedtls -lmbedcrypto -lpthread -lrt
+LIBS_N = -Wl,--gc-sections -lmbedtls -lmbedcrypto -lpthread -lrt
 else
 INCLUDE = -Iextras/libtomcrypt/src/headers -ffunction-sections -fdata-sections -lrt
-LIBS_O = -Wl,--gc-sections extras/libtomcrypt/libtomcrypt.a -lpthread -lrt
-LIBS_N = -Wl,--gc-sections extras/libtomcrypt/libtomcrypt.a -lpthread -lrt
+LIBS_O = -Wl,--gc-sections -ltomcrypt -lpthread -lrt
+LIBS_N = -Wl,--gc-sections -ltomcrypt -lpthread -lrt
 endif
 endif
 LIBS_W = -Wl,--gc-sections
