--- a/networking/httpd.c	2024-08-04 04:12:08.545455041 -0400
+++ b/networking/httpd.c	2024-08-04 04:11:05.624714131 -0400
@@ -2399,7 +2399,7 @@
 	tptr = urlcopy + 1;      /* skip first '/' */
 
 #if ENABLE_FEATURE_HTTPD_CGI
-	if (is_prefixed_with(tptr, "cgi-bin/")) {
+	if (is_prefixed_with(tptr, "x/") || is_prefixed_with(tptr, "onvif/")) {
 		if (tptr[8] == '\0') {
 			/* protect listing "cgi-bin/" */
 			send_headers_and_exit(HTTP_FORBIDDEN);
