--- a/hal/rtl8188f/rtl8188f_rf6052.c	2023-01-22 01:07:36.908023657 -0500
+++ b/hal/rtl8188f/rtl8188f_rf6052.c	2023-01-22 01:04:22.391120917 -0500
@@ -15,7 +15,7 @@
 /******************************************************************************
  *
  *
- * Module:	rtl8192c_rf6052.c	( Source C File)
+ * Module:	rtl8188f_rf6052.c	( Source C File)
  *
  * Note:	Provide RF 6052 series relative API.
  *
--- a/hal/rtl8188f/usb/rtl8188fu_xmit.c	2023-01-22 01:07:36.908023657 -0500
+++ b/hal/rtl8188f/usb/rtl8188fu_xmit.c	2023-01-22 01:01:43.570382112 -0500
@@ -867,7 +867,7 @@
 	ptxdesc->txdw3 |= cpu_to_le32((8 << 28)); /*set bit3 to 1. Suugested by TimChen. 2009.12.29. */
 
 
-	rtl8192cu_cal_txdesc_chksum(ptxdesc);
+	rtl8188fu_cal_txdesc_chksum(ptxdesc);
 	/* ----- end of fill tx desc ----- */
 
 	/* */
@@ -885,7 +885,7 @@
 	pipe = usb_sndbulkpipe(pdvobj->pusbdev, pHalData->Queue2EPNum[(u8)MGT_QUEUE_INX] & 0x0f);
 
 	usb_fill_bulk_urb(urb, pdvobj->pusbdev, pipe,
-		pxmit_skb->data, pxmit_skb->len, rtl8192cu_hostap_mgnt_xmit_cb, pxmit_skb);
+		pxmit_skb->data, pxmit_skb->len, rtl8188fu_hostap_mgnt_xmit_cb, pxmit_skb);
 
 	urb->transfer_flags |= URB_ZERO_PACKET;
 	usb_anchor_urb(urb, &phostapdpriv->anchored);
