--- a/tests/example-deconstructor-alt.c
+++ b/tests/example-deconstructor-alt.c
@@ -714,14 +714,14 @@
 
         case YAML_SCANNER_ERROR:
             if (parser.context) {
-                fprintf(stderr, "Scanner error: %s at line %lu, column %lu\n"
+                fprintf(stderr, "Scanner error: %s at line %u, column %u\n"
                         "%s at line %lu, column %lu\n", parser.context,
                         parser.context_mark.line+1, parser.context_mark.column+1,
                         parser.problem, parser.problem_mark.line+1,
                         parser.problem_mark.column+1);
             }
             else {
-                fprintf(stderr, "Scanner error: %s at line %lu, column %lu\n",
+                fprintf(stderr, "Scanner error: %s at line %u, column %u\n",
                         parser.problem, parser.problem_mark.line+1,
                         parser.problem_mark.column+1);
             }
@@ -729,14 +729,14 @@
 
         case YAML_PARSER_ERROR:
             if (parser.context) {
-                fprintf(stderr, "Parser error: %s at line %lu, column %lu\n"
+                fprintf(stderr, "Parser error: %s at line %u, column %u\n"
                         "%s at line %lu, column %lu\n", parser.context,
                         parser.context_mark.line+1, parser.context_mark.column+1,
                         parser.problem, parser.problem_mark.line+1,
                         parser.problem_mark.column+1);
             }
             else {
-                fprintf(stderr, "Parser error: %s at line %lu, column %lu\n",
+                fprintf(stderr, "Parser error: %s at line %u, column %u\n",
                         parser.problem, parser.problem_mark.line+1,
                         parser.problem_mark.column+1);
             }
