
--- /dev/null
+++ b/arch/mips/xburst/soc-t15/chip-t15/bootes/common/Makefile
@@ -0,0 +1,21 @@
+
+obj-y		+= board_base.o
+obj-y		+= misc.o
+obj-y		+= mmc.o
+obj-$(CONFIG_KEYBOARD_GPIO)	+= keyboard_gpio.o
+obj-$(CONFIG_REGULATOR_DA9024) 	+= regulator.o
+
+
+obj-$(CONFIG_BCM_PM_CORE)   += bcm_power_control.o
+obj-$(CONFIG_VIDEO_TX_ISP)	+= isp_devices.o
+obj-$(CONFIG_SOUND_OSS_XBURST)		+= sound.o
+obj-$(CONFIG_I2C_GPIO)			+= i2c_bus.o
+obj-$(CONFIG_I2C0_V12_JZ)		+= i2c_bus.o
+obj-$(CONFIG_I2C1_V12_JZ)		+= i2c_bus.o
+obj-$(CONFIG_I2C2_V12_JZ)		+= i2c_bus.o
+obj-$(CONFIG_I2C3_V12_JZ)		+= i2c_bus.o
+obj-$(CONFIG_SPI_GPIO)			+= spi_bus.o
+obj-$(CONFIG_SPI0_V12_JZ)		+= spi_bus.o
+obj-$(CONFIG_SPI1_V12_JZ)		+= spi_bus.o
+obj-$(CONFIG_JZ_VO)				+= vo.o
+obj-$(CONFIG_MTD_JZ_SFC_NORFLASH)	+= spi_bus.o
