
--- /dev/null
+++ b/arch/mips/xburst/soc-4780/board/Makefile
@@ -0,0 +1,34 @@
+
+
+obj-$(CONFIG_BOARD_PISCES)		+= pisces/
+obj-$(CONFIG_BOARD_TEST)		+= test/
+obj-$(CONFIG_BOARD_GIANT)		+= giant/
+obj-$(CONFIG_BOARD_WARRIOR)		+= warrior/
+obj-$(CONFIG_BOARD_WARRIOR_V_1_1)	+= warrior/
+obj-$(CONFIG_BOARD_URBOARD)		+= urboard/
+obj-$(CONFIG_BOARD_TROOPER)	        += warrior/
+obj-$(CONFIG_BOARD_PRINTER)	        += printer/
+obj-$(CONFIG_BOARD_HDMI_80)		+= hdmi-80/
+obj-$(CONFIG_BOARD_HLBA1)		+= hlba1/
+obj-$(CONFIG_BOARD_M80)			+= m80/
+obj-$(CONFIG_BOARD_M80B)		+= m80/
+obj-$(CONFIG_BOARD_M80701)		+= m80701/
+obj-$(CONFIG_BOARD_CUCKOO)	        += npm801/
+obj-$(CONFIG_BOARD_NPM801)		+= npm801/
+obj-$(CONFIG_BOARD_T700D)		+= t700d/
+obj-$(CONFIG_BOARD_LEAF)		+= leaf/
+obj-$(CONFIG_BOARD_JI8070A)		+= ji8070a/
+obj-$(CONFIG_BOARD_JI8070B)		+= ji8070a/
+obj-$(CONFIG_BOARD_Q8)			+= ji8070a/
+obj-$(CONFIG_BOARD_I88)			+= i88/
+obj-$(CONFIG_BOARD_GRUS)		+= grus/
+obj-$(CONFIG_BOARD_GRUS_V_1_0_1)	+= grus/
+obj-$(CONFIG_BOARD_NPM3701)		+= npm3701/
+obj-$(CONFIG_BOARD_ZPAD80)		+= zpad80/
+obj-$(CONFIG_BOARD_VGA_DONGLE)		+= vga-dongle/
+obj-$(CONFIG_BOARD_EBOOK)		+= ebook/
+obj-$(CONFIG_BOARD_NC750)		+= nc750/
+obj-$(CONFIG_BOARD_VEHICLE)		+= vehicle/
+obj-$(CONFIG_BOARD_NPM709J)             += npm709j/
+obj-$(CONFIG_BOARD_RY1000)     += ry1000/
+obj-$(CONFIG_BOARD_DS_J801MB)		+= ds-j801mb/
