
--- /dev/null
+++ b/arch/mips/xburst/soc-t30/chip-t30/isvp/common/Makefile
@@ -0,0 +1,12 @@
+
+obj-y		+= board_base.o
+obj-y		+= misc.o
+obj-y		+= mmc.o
+obj-$(CONFIG_KEYBOARD_GPIO)	+= keyboard_gpio.o
+
+obj-$(CONFIG_BCM_PM_CORE)   += bcm_power_control.o
+obj-$(CONFIG_SOUND_OSS_XBURST)		+= sound.o
+obj-$(CONFIG_I2C_GPIO)			+= i2c_bus.o
+obj-$(CONFIG_I2C0_V12_JZ)		+= i2c_bus.o
+obj-$(CONFIG_I2C1_V12_JZ)		+= i2c_bus.o
+obj-$(CONFIG_MTD_JZ_SFC_NORFLASH)	+= spi_bus.o
\ No newline at end of file
