
--- /dev/null
+++ b/arch/mips/xburst/soc-m200/Kconfig
@@ -0,0 +1,30 @@
+choice
+	prompt "PRODUCT type"
+	depends on SOC_M200
+	default PRODUCT_DEVELOP
+
+config BOARD_4785_FPGA
+       bool "FPGA"
+
+config PRODUCT_DEVELOP
+       bool "DEVELOP"
+
+config BOARD_NEWTON
+       bool "NEWTON"
+config BOARD_FORNAX
+       bool "FORNAX"
+endchoice
+
+config PRODUCT_NAME
+	string
+	default "fpga" if BOARD_4785_FPGA
+	default "dorado" if PRODUCT_DEVELOP
+	default "newton" if BOARD_NEWTON
+	default "fornax" if BOARD_FORNAX
+
+if PRODUCT_DEVELOP
+source "arch/mips/xburst/soc-m200/board/dorado/Kconfig"
+endif
+if BOARD_NEWTON
+source "arch/mips/xburst/soc-m200/board/newton/Kconfig"
+endif
