
--- /dev/null
+++ b/arch/mips/xburst/soc-m200/board/dorado/Kconfig
@@ -0,0 +1,38 @@
+choice
+	prompt "BOARD type"
+	depends on PRODUCT_DEVELOP
+	default BOARD_DORADO_V20
+
+config BOARD_DORADO_V20
+       bool "board dorado v20"
+       depends on PRODUCT_DEVELOP
+
+config BOARD_DORADO_V21
+       bool "board dorado v21"
+       depends on PRODUCT_DEVELOP
+
+config BOARD_DORADO_V22
+       bool "board dorado v22"
+       depends on PRODUCT_DEVELOP
+
+config BOARD_DORADO_V30
+       bool "board dorado v30"
+       depends on PRODUCT_DEVELOP
+
+endchoice
+
+config BOARD_NAME
+	string
+	default "dorado_v20" if BOARD_DORADO_V20
+
+config BOARD_NAME
+	string
+	default "dorado_v21" if BOARD_DORADO_V21
+
+config BOARD_NAME
+	string
+	default "dorado_v22" if BOARD_DORADO_V22
+
+config BOARD_NAME
+	string
+	default "dorado_v30" if BOARD_DORADO_V30
