
--- a/arch/mips/kernel/process.c
+++ b/arch/mips/kernel/process.c
@@ -42,6 +42,8 @@
 #include <asm/inst.h>
 #include <asm/stacktrace.h>
 
+#include <mxu.h>
+
 #ifdef CONFIG_HOTPLUG_CPU
 void arch_cpu_idle_dead(void)
 {
@@ -60,15 +62,14 @@ void start_thread(struct pt_regs * regs, unsigned long pc, unsigned long sp)
 
 	/* New thread loses kernel privileges. */
 	status = regs->cp0_status & ~(ST0_CU0|ST0_CU1|ST0_FR|KU_MASK);
-#ifdef CONFIG_64BIT
-	status |= test_thread_flag(TIF_32BIT_REGS) ? 0 : ST0_FR;
-#endif
 	status |= KU_USER;
 	regs->cp0_status = status;
 	clear_used_math();
 	clear_fpu_owner();
 	if (cpu_has_dsp)
 		__init_dsp();
+	if (cpu_has_mxu)
+		__init_mxu();
 	regs->cp0_epc = pc;
 	regs->regs[29] = sp;
 }
@@ -99,6 +100,9 @@ int copy_thread(unsigned long clone_flags, unsigned long usp,
 	if (cpu_has_dsp)
 		save_dsp(p);
 
+	if (cpu_has_mxu)
+		save_mxu(p);
+
 	preempt_enable();
 
 	/* set up new TSS. */
