
--- /dev/null
+++ b/arch/mips/xburst/soc-4780/board/giant/giant-misc.c
@@ -0,0 +1,190 @@
+#include <linux/platform_device.h>
+#include <linux/i2c.h>
+#include <linux/gpio_keys.h>
+#include <linux/input.h>
+#include <mach/platform.h>
+#include <mach/jzsnd.h>
+#include <mach/jzmmc.h>
+#include <gpio.h>
+
+#include "giant.h"
+
+#ifdef CONFIG_KEYBOARD_GPIO
+static struct gpio_keys_button board_buttons[] = {
+	{
+		.gpio		= GPIO_PF(29),
+		.code   	= KEY_HOME,
+		.desc		= "home key",
+		.active_low	= 1,
+	},
+};
+
+static struct gpio_keys_platform_data board_button_data = {
+	.buttons	= board_buttons,
+	.nbuttons	= ARRAY_SIZE(board_buttons),
+};
+
+static struct platform_device jz_button_device = {
+	.name		= "gpio-keys",
+	.id		= -1,
+	.num_resources	= 0,
+	.dev		= {
+		.platform_data	= &board_button_data,
+	}
+};
+#endif
+
+#ifdef CONFIG_JZ4780_SUPPORT_TSC
+static struct jztsc_pin giant_tsc_gpio[] = {
+	[0] = {GPIO_GT801_IRQ,		HIGH_ENABLE},
+	[1] = {GPIO_GT801_SHUTDOWN,	HIGH_ENABLE},
+	[2] = {GPIO_TP_DRV_EN,		HIGH_ENABLE},
+};
+
+static struct jztsc_platform_data giant_tsc_pdata = {
+	.gpio	= giant_tsc_gpio,
+};
+
+static struct i2c_board_info giant_i2c0_devs[] __initdata = {
+	{
+		I2C_BOARD_INFO("gt801_ts", 0x55),
+		.irq = 0,
+		.platform_data	= &giant_tsc_pdata,
+	},
+};
+#endif
+
+static int __init giant_board_init(void)
+{
+/* dma */
+#ifdef CONFIG_XBURST_DMAC
+	platform_device_register(&jz_pdma_device);
+#endif
+/* i2c */
+#ifdef CONFIG_I2C0_JZ4780
+	platform_device_register(&jz_i2c0_device);
+#endif
+#ifdef CONFIG_I2C1_JZ4780
+	platform_device_register(&jz_i2c1_device);
+#endif
+#ifdef CONFIG_I2C2_JZ4780
+	platform_device_register(&jz_i2c2_device);
+#endif
+#ifdef CONFIG_I2C3_JZ4780
+	platform_device_register(&jz_i2c3_device);
+#endif
+#ifdef CONFIG_I2C4_JZ4780
+	platform_device_register(&jz_i2c4_device);
+#endif
+
+/* ipu */
+#ifdef CONFIG_JZ4780_IPU
+	platform_device_register(&jz_ipu0_device);
+#endif
+#ifdef CONFIG_JZ4780_IPU
+	platform_device_register(&jz_ipu1_device);
+#endif
+
+/* mmc */
+#ifdef CONFIG_MMC0_JZ4780
+	jz_device_register(&jz_msc0_device, &giant_inand_pdata);
+#endif
+#ifdef CONFIG_MMC1_JZ4780
+	jz_device_register(&jz_msc1_device, &giant_tf_pdata);
+#endif
+#ifdef CONFIG_MMC2_JZ4780
+	jz_device_register(&jz_msc2_device, &giant_sdio_pdata);
+#endif
+
+/* sound */
+#ifdef CONFIG_SOUND_I2S0_JZ47XX
+	jz_device_register(&jz_i2s0_device,&i2s0_data);
+#endif
+#ifdef CONFIG_SOUND_I2S1_JZ47XX
+	jz_device_register(&jz_i2s1_device,&i2s1_data);
+#endif
+#ifdef CONFIG_SOUND_PCM0_JZ47XX
+	jz_device_register(&jz_pcm0_device,&pcm0_data);
+#endif
+#ifdef CONFIG_SOUND_PCM1_JZ47XX
+	jz_device_register(&jz_pcm1_device,&pcm1_data);
+#endif
+#ifdef CONFIG_JZ4780_INTERNAL_CODEC
+	jz_device_register(&jz_codec_device, &codec_data);
+#endif
+
+/* panel and bl */
+#ifdef CONFIG_LCD_AUO_A043FL01V2
+	platform_device_register(&auo_a043fl01v2_device);
+#endif
+#ifdef CONFIG_LCD_AT070TN93
+	platform_device_register(&at070tn93_device);
+#endif
+#ifdef CONFIG_BACKLIGHT_PWM
+	platform_device_register(&giant_backlight_device);
+#endif
+
+/* lcdc framebuffer*/
+#ifdef CONFIG_FB_JZ4780_LCDC1
+	jz_device_register(&jz_fb1_device, &jzfb1_pdata);
+#endif
+#ifdef CONFIG_FB_JZ4780_LCDC0
+	jz_device_register(&jz_fb0_device, &jzfb0_pdata);
+#endif
+
+/* uart */
+#ifdef CONFIG_SERIAL_JZ47XX_UART0
+	platform_device_register(&jz_uart0_device);
+#endif
+#ifdef CONFIG_SERIAL_JZ47XX_UART1
+	platform_device_register(&jz_uart1_device);
+#endif
+#ifdef CONFIG_SERIAL_JZ47XX_UART2
+	platform_device_register(&jz_uart2_device);
+#endif
+#ifdef CONFIG_SERIAL_JZ47XX_UART3
+	platform_device_register(&jz_uart3_device);
+#endif
+#ifdef CONFIG_SERIAL_JZ47XX_UART4
+	platform_device_register(&jz_uart4_device);
+#endif
+
+#ifdef CONFIG_JZ_CIM
+	platform_device_register(&jz_cim_device);
+#endif
+
+#ifdef CONFIG_USB_OHCI_HCD
+	platform_device_register(&jz_ohci_device);
+#endif
+
+#ifdef CONFIG_USB_EHCI_HCD
+	platform_device_register(&jz_ehci_device);
+#endif
+
+#ifdef CONFIG_JZ_MAC
+	platform_device_register(&jz_mac);
+#endif
+
+#ifdef CONFIG_KEYBOARD_GPIO
+	platform_device_register(&jz_button_device);
+#endif
+/* nand */
+#ifdef CONFIG_NAND_DRIVER
+	jz_device_register(&jz_nand_device, NULL);
+#endif
+
+#ifdef CONFIG_HDMI_JZ4780
+	platform_device_register(&jz_hdmi);
+#endif
+
+#ifdef CONFIG_JZ4780_SUPPORT_TSC
+	i2c_register_board_info(0, giant_i2c0_devs, ARRAY_SIZE(giant_i2c0_devs));
+#endif	
+#ifdef CONFIG_RTC_DRV_JZ4780
+	platform_device_register(&jz_rtc_device);
+#endif
+	return 0;
+}
+
+
+arch_initcall(giant_board_init);
