
--- /dev/null
+++ b/arch/mips/xburst/soc-t15/chip-t15/Kconfig
@@ -0,0 +1,42 @@
+
+choice
+	prompt "Platform type"
+	depends on CHIP_T15
+	default BOARD_FPGA_T15
+
+config BOARD_FPGA_T15
+	bool "FPGA"
+
+config BOARD_HERCULES
+	bool "Hercules"
+
+config BOARD_BOX
+	bool "box"
+
+config BOARD_BOOTES
+	bool "Bootes"
+
+endchoice
+
+config PRODUCT_NAME
+	string
+	default "fpga" if BOARD_FPGA_T15
+	default "hercules" if BOARD_HERCULES
+	default "box" if BOARD_BOX
+	default "bootes" if BOARD_BOOTES
+
+if BOARD_FPGA_T15
+source "arch/mips/xburst/soc-t15/chip-t15/fpga/Kconfig"
+endif
+
+if BOARD_BOX
+source "arch/mips/xburst/soc-t15/chip-t15/box/Kconfig"
+endif
+
+if BOARD_HERCULES
+source "arch/mips/xburst/soc-t15/chip-t15/hercules/Kconfig"
+endif
+
+if BOARD_BOOTES
+source "arch/mips/xburst/soc-t15/chip-t15/bootes/Kconfig"
+endif
