
--- /dev/null
+++ b/arch/mips/xburst/soc-4780/board/hdmi-80/Makefile
@@ -0,0 +1,8 @@
+
+obj-y += hdmi-80-misc.o hdmi-80-pm.o hdmi-80-lcd.o hdmi-80-regulator.o  
+
+obj-$(CONFIG_I2C)		+= hdmi-80-i2c.o
+#obj-$(CONFIG_MFD_ACT8600)	+= hdmi-80-regulator.o
+obj-$(CONFIG_MMC_JZ4780)	+= hdmi-80-mmc.o
+obj-$(CONFIG_JZ4780_INTERNAL_CODEC) += hdmi-80-sound.o
+obj-$(CONFIG_MTD_NAND_JZ4780) += hdmi-80-nand-mtd.o
