
--- /dev/null
+++ b/arch/mips/xburst/soc-t15/chip-t15/fpga/common/lcd/Makefile
@@ -0,0 +1,10 @@
+
+obj-$(CONFIG_LCD_TRULY_TFT240240_2_E) += lcd-truly_tft240240_2_e.o
+obj-$(CONFIG_LCD_BYD_8991FTGF) += lcd-8991.o
+obj-$(CONFIG_LCD_CV90_M5377_P30) += lcd-mirasol.o
+obj-$(CONFIG_LCD_OTM8012A_TFT1P6449) += lcd-otm8012a.o
+obj-$(CONFIG_LCD_BYD_9177AA) += lcd-9177aa.o
+obj-$(CONFIG_LCD_TRULY_TDO_HD0499K) += lcd-truly_tdo_hd0499k.o
+obj-$(CONFIG_LCD_LH155) += lcd-lh155.o
+obj-$(CONFIG_JZ_EPD_V12)	+= epd.o
+obj-$(CONFIG_LCD_BYD_BM8766U) += lcd-8766.o
