
--- /dev/null
+++ b/arch/mips/xburst/soc-4780/board/ji8070a/Makefile
@@ -0,0 +1,9 @@
+
+obj-y += ji8070a-misc.o ji8070a-pm.o ji8070a-lcd.o  
+
+obj-$(CONFIG_I2C)		+= ji8070a-i2c.o
+obj-$(CONFIG_MFD_ACT8600)	+= ji8070a-regulator.o
+obj-$(CONFIG_MMC_JZ4780)	+= ji8070a-mmc.o
+obj-$(CONFIG_JZ4780_INTERNAL_CODEC) += ji8070a-sound.o
+obj-$(CONFIG_VIDEO_JZ4780_CIM_HOST)	+= ji8070a-cim.o
+obj-$(CONFIG_MTD_NAND_JZ4780) += ji8070a-nand-mtd.o
