
--- /dev/null
+++ b/arch/mips/xburst/soc-t15/common/Makefile
@@ -0,0 +1,27 @@
+
+
+## WARNING! DO NOT change this lists order because power sleep order depends on it!! ##
+
+obj-y += setup.o
+obj-y += reset.o
+obj-y += timer_sys_ost.o
+#obj-y += timer_tcu.o
+#maintimer.o cpu0timer.o cpu1timer.o
+obj-y += serial.o
+
+obj-y += clk/
+
+obj-y += gpio.o
+# only for debug
+#obj-y += pmmon.o
+obj-y += irq.o
+obj-y += cpm_usb.o
+
+obj-y += platform.o
+
+#obj-y += pm.o
+
+#obj-y += ahbm.o
+
+obj-y += libdmmu.o
+obj-y += ddr.o
