
--- a/arch/mips/include/asm/asmmacro-64.h
+++ b/arch/mips/include/asm/asmmacro-64.h
@@ -54,7 +54,7 @@
 	.endm
 
 	.macro	fpu_save_double thread status tmp
-	sll	\tmp, \status, 5
+	sll     \tmp, \status, 31 - _ST0_FR
 	bgez	\tmp, 2f
 	fpu_save_16odd \thread
 2:
@@ -102,7 +102,7 @@
 	.endm
 
 	.macro	fpu_restore_double thread status tmp
-	sll	\tmp, \status, 5
+	sll     \tmp, \status, 31 - _ST0_FR
 	bgez	\tmp, 1f				# 16 register mode?
 
 	fpu_restore_16odd \thread
