
--- /dev/null
+++ b/arch/mips/xburst/Makefile
@@ -0,0 +1,13 @@
+
+obj-y			+= core/
+obj-y			+= common/
+
+obj-$(CONFIG_SOC_4775) 	+= soc-4775/
+obj-$(CONFIG_SOC_4780) 	+= soc-4780/
+obj-$(CONFIG_SOC_M200) 	+= soc-m200/
+obj-$(CONFIG_SOC_T15) 	+= soc-t15/
+obj-$(CONFIG_SOC_T10) 	+= soc-t10/
+obj-$(CONFIG_SOC_T20) 	+= soc-t20/
+obj-$(CONFIG_SOC_T30) 	+= soc-t30/
+obj-$(CONFIG_SOC_T21) 	+= soc-t21/
+obj-$(CONFIG_SOC_T31) 	+= soc-t31/
