
--- /dev/null
+++ b/arch/mips/xburst/soc-4775/chip-4775/Kconfig
@@ -0,0 +1,39 @@
+
+
+choice
+	prompt "Board type"
+	depends on SOC_4775
+	default BOARD_4775_TEST
+
+config BOARD_4775_ORION
+        bool "ORION"
+
+config BOARD_4775_MENSA
+        bool "MENSA"
+
+config BOARD_4775_COMET
+        bool "COMET"
+
+config BOARD_4775_CONE
+	bool "CONE"
+
+config BOARD_4775_CONE_V1_2
+        bool "CONE_V1_2"
+
+config BOARD_4775_HSUN
+	bool "HSUN"
+
+config BOARD_4775_SEDBIO
+	bool "SEDBIO"
+
+endchoice
+
+config BOARD_NAME
+	string
+	default "orion" if BOARD_4775_ORION
+	default "mensa" if BOARD_4775_MENSA
+	default "comet" if BOARD_4775_COMET
+	default "cone" if BOARD_4775_CONE
+	default "cone_v1_2" if BOARD_4775_CONE_V1_2
+	default "hsun" if BOARD_4775_HSUN
+	default "sedbio" if BOARD_4775_SEDBIO
