
--- /dev/null
+++ b/arch/mips/xburst/soc-t10/chip-t10/Kconfig
@@ -0,0 +1,27 @@
+
+choice
+	prompt "Platform type"
+	depends on CHIP_T10
+	default PLATFORM_ISVP
+
+config BOARD_MUSCA
+	bool "Musca"
+
+config PLATFORM_ISVP
+	bool "isvp"
+
+endchoice
+
+config PRODUCT_NAME
+	string
+	default "musca" if BOARD_MUSCA
+	default "isvp" if PLATFORM_ISVP
+
+
+if BOARD_MUSCA
+source "arch/mips/xburst/soc-t10/chip-t10/musca/Kconfig"
+endif
+
+if PLATFORM_ISVP
+source "arch/mips/xburst/soc-t10/chip-t10/isvp/Kconfig"
+endif
