
--- a/arch/mips/include/asm/cop2.h
+++ b/arch/mips/include/asm/cop2.h
@@ -11,6 +11,24 @@
 
 #include <linux/notifier.h>
 
+#if defined(CONFIG_MACH_XBURST)
+
+extern void xburst_cop2_save(struct xburst_cop2_state *);
+extern void xburst_cop2_restore(struct xburst_cop2_state *);
+#define cop2_save(r)		xburst_cop2_save(r)
+#define cop2_restore(r)		xburst_cop2_restore(r)
+
+#define cop2_present		1
+#define cop2_lazy_restore	0
+
+#else
+
+#define cop2_present		0
+#define cop2_lazy_restore	0
+#define cop2_save(r)
+#define cop2_restore(r)
+#endif
+
 enum cu2_ops {
 	CU2_EXCEPTION,
 	CU2_LWC2_OP,
