
--- /dev/null
+++ b/arch/mips/xburst/Platform
@@ -0,0 +1,14 @@
+
+platform-$(CONFIG_MACH_XBURST)	+= xburst/
+cflags-$(CONFIG_MACH_XBURST)	+= -I$(srctree)/arch/mips/xburst/core/include
+load-$(CONFIG_MACH_XBURST)	+= 0xffffffff80010000
+
+cflags-$(CONFIG_SOC_4775)	+= -I$(srctree)/arch/mips/xburst/soc-4775/include
+cflags-$(CONFIG_SOC_4780)	+= -I$(srctree)/arch/mips/xburst/soc-4780/include
+cflags-$(CONFIG_SOC_M200)	+= -I$(srctree)/arch/mips/xburst/soc-m200/include
+cflags-$(CONFIG_SOC_T15)	+= -I$(srctree)/arch/mips/xburst/soc-t15/include
+cflags-$(CONFIG_SOC_T10)	+= -I$(srctree)/arch/mips/xburst/soc-t10/include
+cflags-$(CONFIG_SOC_T20)	+= -I$(srctree)/arch/mips/xburst/soc-t20/include
+cflags-$(CONFIG_SOC_T30)	+= -I$(srctree)/arch/mips/xburst/soc-t30/include
+cflags-$(CONFIG_SOC_T21)	+= -I$(srctree)/arch/mips/xburst/soc-t21/include
+cflags-$(CONFIG_SOC_T31)	+= -I$(srctree)/arch/mips/xburst/soc-t31/include
