
--- a/arch/mips/math-emu/ieee754int.h
+++ b/arch/mips/math-emu/ieee754int.h
@@ -83,10 +83,17 @@
     if(ve == SP_EMAX+1+SP_EBIAS){\
 	if(vm == 0)\
 	  vc = IEEE754_CLASS_INF;\
-	else if(vm & SP_MBIT(SP_MBITS-1)) \
-	  vc = IEEE754_CLASS_SNAN;\
-	else \
-	  vc = IEEE754_CLASS_QNAN;\
+	else if (ieee754_csr.nan2008) { \
+	  if(vm & SP_MBIT(SP_MBITS-1)) \
+	    vc = IEEE754_CLASS_QNAN;\
+	  else \
+	    vc = IEEE754_CLASS_SNAN;\
+	} else { \
+	  if(vm & SP_MBIT(SP_MBITS-1)) \
+	    vc = IEEE754_CLASS_SNAN;\
+	  else \
+	    vc = IEEE754_CLASS_QNAN;\
+	} \
     } else if(ve == SP_EMIN-1+SP_EBIAS) {\
 	if(vm) {\
 	    ve = SP_EMIN;\
@@ -117,10 +124,17 @@ u64 ym; int ye; int ys; int yc
     if(ve == DP_EMAX+1+DP_EBIAS){\
 	if(vm == 0)\
 	  vc = IEEE754_CLASS_INF;\
-	else if(vm & DP_MBIT(DP_MBITS-1)) \
-	  vc = IEEE754_CLASS_SNAN;\
-	else \
-	  vc = IEEE754_CLASS_QNAN;\
+	else if (ieee754_csr.nan2008) { \
+	  if(vm & DP_MBIT(DP_MBITS-1)) \
+	    vc = IEEE754_CLASS_QNAN;\
+	  else \
+	    vc = IEEE754_CLASS_SNAN;\
+	} else { \
+	  if(vm & DP_MBIT(DP_MBITS-1)) \
+	    vc = IEEE754_CLASS_SNAN;\
+	  else \
+	    vc = IEEE754_CLASS_QNAN;\
+	} \
     } else if(ve == DP_EMIN-1+DP_EBIAS) {\
 	if(vm) {\
 	    ve = DP_EMIN;\
