
--- /dev/null
+++ b/arch/mips/xburst/soc-4780/board/npm801/Makefile
@@ -0,0 +1,9 @@
+
+obj-y += npm801-misc.o npm801-pm.o npm801-lcd.o  
+
+obj-$(CONFIG_I2C)		+= npm801-i2c.o
+obj-$(CONFIG_MFD_ACT8600)	+= npm801-regulator.o
+obj-$(CONFIG_MMC_JZ4780)	+= npm801-mmc.o
+obj-$(CONFIG_JZ4780_INTERNAL_CODEC) += npm801-sound.o
+obj-$(CONFIG_JZ_MODEM)		+= npm801-modem.o
+obj-$(CONFIG_BCM4330_RFKILL) += npm801-bt.o
