
--- a/arch/mips/kernel/genex.S
+++ b/arch/mips/kernel/genex.S
@@ -374,10 +374,26 @@ NESTED(except_vec_nmi, 0, sp)
 NESTED(nmi_handler, PT_SIZE, sp)
 	.set	push
 	.set	noat
+		/* Clear ERL - restore segment mapping */
+		mfc0	k0, CP0_STATUS
+		ori     k0, k0, ST0_EXL
+		lui     k1, 0xffff & ~(ST0_BEV>>16)
+		ori     k1, k1, 0xffff & ~(ST0_ERL)
+		and     k0, k0, k1
+		mtc0    k0, CP0_STATUS
+		ehb
 	SAVE_ALL
 	move	a0, sp
 	jal	nmi_exception_handler
 	RESTORE_ALL
+		/* Set ERL and clear EXL|NMI */
+		mfc0	k0, CP0_STATUS
+		ori     k0, k0, ST0_ERL
+		lui     k1, 0xffff & ~(ST0_NMI>>16)
+		ori     k1, k1, 0xffff & ~(ST0_EXL)
+		and     k0, k0, k1
+		mtc0    k0, CP0_STATUS
+		ehb
 	.set	mips3
 	eret
 	.set	pop
@@ -468,6 +484,7 @@ NESTED(nmi_handler, PT_SIZE, sp)
 	BUILD_HANDLER ov ov sti silent			/* #12 */
 	BUILD_HANDLER tr tr sti silent			/* #13 */
 	BUILD_HANDLER fpe fpe fpe silent		/* #15 */
+	BUILD_HANDLER mfpe mfpe sti silent		/* #16 */
 	BUILD_HANDLER mdmx mdmx sti silent		/* #22 */
 #ifdef	CONFIG_HARDWARE_WATCHPOINTS
 	/*
