
--- /dev/null
+++ b/arch/mips/xburst/soc-4780/board/npm709j/Makefile
@@ -0,0 +1,10 @@
+
+obj-y += npm709j-misc.o npm709j-pm.o npm709j-lcd.o  
+
+obj-$(CONFIG_I2C)		+= npm709j-i2c.o
+obj-$(CONFIG_MFD_ACT8600)	+= npm709j-regulator.o
+obj-$(CONFIG_MMC_JZ4780)	+= npm709j-mmc.o
+obj-$(CONFIG_JZ4780_INTERNAL_CODEC) += npm709j-sound.o
+obj-$(CONFIG_JZ_MODEM)		+= npm709j-modem.o
+obj-$(CONFIG_BCM4330_RFKILL) += npm709j-bt.o
+obj-$(CONFIG_VIDEO_JZ4780_CIM_HOST) += npm709j-cim.o
