
--- /dev/null
+++ b/arch/mips/xburst/soc-m200/board/fpga/mmc.c
@@ -0,0 +1,62 @@
+#include <linux/mmc/host.h>
+#include <linux/regulator/consumer.h>
+#include <linux/gpio.h>
+#include <linux/wakelock.h>
+#include <linux/err.h>
+#include <linux/delay.h>
+
+#include <mach/jzmmc.h>
+
+#include "board.h"
+
+
+#ifndef CONFIG_NAND
+#ifdef CONFIG_JZMMC_V12_MMC0
+
+/* on fpga 4785, use MMC_CAP_1_BIT_DATA instead of MMC_CAP_4_BIT_DATA */
+struct jzmmc_platform_data inand_pdata = {
+	.removal  			= DONTCARE,
+	.sdio_clk			= 0,
+	.ocr_avail			= MMC_VDD_32_33 | MMC_VDD_33_34,
+	.capacity  			= MMC_CAP_SD_HIGHSPEED | MMC_CAP_MMC_HIGHSPEED /*| MMC_CAP_4_BIT_DATA */| MMC_CAP_NONREMOVABLE,
+	.pm_flags			= 0,
+	.max_freq			= CONFIG_MMC0_MAX_FREQ,
+	.gpio				= NULL,
+#ifdef CONFIG_MMC0_PIO_MODE
+	.pio_mode			= 1,
+#else
+	.pio_mode			= 0,
+#endif
+	.private_init			= NULL,
+};
+#endif
+
+#ifdef CONFIG_JZMMC_V12_MMC1
+static struct card_gpio tf_gpio = {
+	.cd				= {-1,	-1},
+	.wp             = {-1,	-1},
+	.pwr			= {-1,	-1},
+};
+
+/* common pdata for both tf_card and sdio wifi on fpga board */
+struct jzmmc_platform_data tf_pdata = {
+	.removal  			= DONTCARE,/*REMOVABLE*/
+	.sdio_clk			= 0,
+	.ocr_avail			= MMC_VDD_32_33 | MMC_VDD_33_34,
+	.capacity  			= MMC_CAP_SD_HIGHSPEED | MMC_CAP_MMC_HIGHSPEED /*| MMC_CAP_4_BIT_DATA*/,
+	.pm_flags			= 0,
+	.recovery_info			= NULL,
+	.gpio				= &tf_gpio,
+	.max_freq                       = CONFIG_MMC1_MAX_FREQ,
+#ifdef CONFIG_MMC1_PIO_MODE
+	.pio_mode                       = 1,
+#else
+	.pio_mode                       = 0,
+#endif
+};
+#endif
+
+#else /* CONFIG NAND*/
+
+
+#endif
