
--- a/drivers/input/touchscreen/Makefile
+++ b/drivers/input/touchscreen/Makefile
@@ -6,6 +6,26 @@
 
 wm97xx-ts-y := wm97xx-core.o
 
+##########-ingenic-touchscreen-###########
+
+obj-$(CONFIG_TOUCHSCREEN_JZ_GT801)	+= tsc_gt801.o
+obj-$(CONFIG_TOUCHSCREEN_LDWZIC)        += ldwzic_ts.o
+obj-$(CONFIG_TOUCHSCREEN_JZ_SSD253X)	+= ssd253x-ts.o
+obj-$(CONFIG_TOUCHSCREEN_FT5X0X)        += ft5x0x_ts.o
+obj-$(CONFIG_TOUCHSCREEN_FT5X06)        += ft5x06_ts.o
+obj-$(CONFIG_TOUCHSCREEN_MG8698S)        += mg8698s_ts.o
+obj-$(CONFIG_GSlX680_CAPACITIVE_TOUCHSCREEN)	+= gslX680.o
+obj-$(CONFIG_TOUCHSCREEN_ZET6221)        += zet6221_ts.o zet6221_downloader.o
+obj-$(CONFIG_TOUCHSCREEN_FT5306)        += ft5306_ts.o
+obj-$(CONFIG_TOUCHSCREEN_CTP_IT7260)    += ctp_it7260.o
+obj-$(CONFIG_TOUCHSCREEN_GWTC9XXXB)     += gwtc9xxxb_ts.o
+obj-$(CONFIG_TOUCHSCREEN_NOVATEK)       += novatek_ts.o
+obj-$(CONFIG_TOUCHSCREEN_GOODIX)	+= goodix_ts.o
+obj-$(CONFIG_JZ4780_SUPPORT_RESISTANCE_TSC) +=jz4780_ts.o
+obj-$(CONFIG_TOUCHSCREEN_FT5336)        += ft5336_ts.o
+obj-$(CONFIG_TOUCHSCREEN_ITE7258)        += ite7258_ts.o
+
+##########################################
 obj-$(CONFIG_TOUCHSCREEN_88PM860X)	+= 88pm860x-ts.o
 obj-$(CONFIG_TOUCHSCREEN_AD7877)	+= ad7877.o
 obj-$(CONFIG_TOUCHSCREEN_AD7879)	+= ad7879.o
@@ -64,6 +84,7 @@ obj-$(CONFIG_TOUCHSCREEN_WACOM_W8001)	+= wacom_w8001.o
 obj-$(CONFIG_TOUCHSCREEN_WACOM_I2C)	+= wacom_i2c.o
 obj-$(CONFIG_TOUCHSCREEN_WM831X)	+= wm831x-ts.o
 obj-$(CONFIG_TOUCHSCREEN_WM97XX)	+= wm97xx-ts.o
+obj-$(CONFIG_TOUCHSCREEN_FT6X06)	+= ft6x06_ts.o
 wm97xx-ts-$(CONFIG_TOUCHSCREEN_WM9705)	+= wm9705.o
 wm97xx-ts-$(CONFIG_TOUCHSCREEN_WM9712)	+= wm9712.o
 wm97xx-ts-$(CONFIG_TOUCHSCREEN_WM9713)	+= wm9713.o
