
--- a/drivers/rtc/Kconfig
+++ b/drivers/rtc/Kconfig
@@ -114,6 +114,23 @@ config RTC_INTF_DEV_UIE_EMUL
 	  clock several times per second, please enable this option
 	  only if you know that you really need it.
 
+config RTC_INTF_ALARM
+        bool "Android alarm driver"
+        depends on RTC_CLASS
+        default y
+        help
+          Provides non-wakeup and rtc backed wakeup alarms based on rtc or
+          elapsed realtime, and a non-wakeup alarm on the monotonic clock.
+          Also provides an interface to set the wall time which must be used
+          for elapsed realtime to work.
+
+config RTC_INTF_ALARM_DEV
+    bool "Android alarm device"
+    depends on RTC_INTF_ALARM
+    default y
+    help
+        Exports the alarm interface to user-space.
+
 config RTC_DRV_TEST
 	tristate "Test driver/device"
 	help
@@ -128,6 +145,23 @@ config RTC_DRV_TEST
 	  This driver can also be built as a module. If so, the module
 	  will be called rtc-test.
 
+config RTC_DRV_JZ
+    tristate "JZ RTC"
+        help
+        If you say Y here you will get access to the real time clock
+          built into your JZ CPU.
+
+          To compile this driver as a module, choose M here: the
+          module will be called rtc-jz.
+
+config USE_EXTERNAL_24M_CLOCK
+    bool "use external 24M clock"
+	depends on RTC_DRV_JZ
+    default n
+        help
+        If you say Y here the RTC will use an external 24M clock
+         as the clock source
+
 comment "I2C RTC drivers"
 	depends on I2C
 
