
--- a/drivers/char/Makefile
+++ b/drivers/char/Makefile
@@ -56,9 +56,16 @@ obj-$(CONFIG_PCMCIA)		+= pcmcia/
 obj-$(CONFIG_HANGCHECK_TIMER)	+= hangcheck-timer.o
 obj-$(CONFIG_TCG_TPM)		+= tpm/
 
+obj-$(CONFIG_DCC_TTY)		+= dcc_tty.o
 obj-$(CONFIG_PS3_FLASH)		+= ps3flash.o
 
 obj-$(CONFIG_JS_RTC)		+= js-rtc.o
 js-rtc-y = rtc.o
+obj-$(CONFIG_JZ_VP)     += jz_virtual_privilege.o
 
 obj-$(CONFIG_TILE_SROM)		+= tile-srom.o
+
+obj-$(CONFIG_JZ_DMIC_V12)	+= voice_wakeup/
+obj-$(CONFIG_JZ_DMIC_V12)	+= jz_dmic.o
+obj-$(CONFIG_JZ_DMIC_WAKEUP) += jz_wakeup.o
+obj-$(CONFIG_TEST_SECOND_REFRESH)	+= second_refresh/
