
--- a/drivers/nfc/Kconfig
+++ b/drivers/nfc/Kconfig
@@ -5,6 +5,26 @@
 menu "Near Field Communication (NFC) devices"
 	depends on NFC
 
+
+config BCM2079X_NFC
+         tristate "BCM2079X NFC driver"
+         depends on I2C
+         select CRC_CCITT
+         default n
+         ---help---
+           Say yes if you want BCM2079X Near Field Communication driver.
+           This is for i2c connected version. If unsure, say N here.
+           To compile this driver as a module, choose m here. The module will
+           be called bcm2079x.
+
+ config BCM2079X_NFC_I2C_BUSNUM
+         int "BCM2079x NFC i2c bus number"
+         depends on BCM2079X_NFC
+         default 0
+         help
+	 The bcm2079x nfc uses which one i2c controller, you must check the schematic.
+             The bcm2079x nfc uses which one i2c controller, you must check the schematic.
+
 config NFC_PN533
 	tristate "NXP PN533 USB driver"
 	depends on USB
