
--- /dev/null
+++ b/drivers/video/jz_vpu/Kconfig
@@ -0,0 +1,26 @@
+#
+# jz vpu configuration
+#
+
+menuconfig JZ_VPU
+	bool"JZ VPU driver"
+	default n
+	help
+		VPU support.
+
+choice
+	depends on JZ_VPU
+	prompt "select vpu type"
+	default JZ_VPU_V12
+
+config JZ_VPU_PRE
+	bool "VPU driver for Ingenic previous version"
+	help
+	  Support for Ingenic VPU muti-thread operations.
+
+config JZ_VPU_V12
+	bool "VPU driver v12 for Ingenic m200"
+	help
+	  Support for Ingenic VPU muti-thread operations.
+
+endchoice
