
--- /dev/null
+++ b/drivers/nand/driver/Kconfig
@@ -0,0 +1,18 @@
+menuconfig NAND
+	tristate "-------- nand driver ---------"
+
+choice
+        prompt "------------ soc select --------------"
+        default NAND_JZ4780
+
+config  NAND_JZ4780
+        prompt "------------ 4780 soc --------------"
+	help
+		help mesage
+
+config  NAND_JZ4790
+        prompt "------------ 4790 soc --------------"
+	help
+		help mesage
+
+endchoice
\ No newline at end of file
