
--- a/drivers/char/Kconfig
+++ b/drivers/char/Kconfig
@@ -6,6 +6,47 @@ menu "Character devices"
 
 source "drivers/tty/Kconfig"
 
+config JZ_VP
+	bool "Virtual Privilege device driver"
+	default n
+	help
+	  Say Y here if you want to support Virtual Privilege device. The
+	  Virtual Privilege device supports flush cache and get phyaddr.
+	  When in doubt, say "N".
+
+config JZ_DMIC_V12
+	bool "Ingenic Dmic Driver"
+	default n
+	help
+		Say Y here if you want to support DMIC on your board. The Dmic Driver
+		support record and wakeup function. When in doubt, Say N here.
+config JZ_DMIC_WAKEUP
+	bool "Ingenic Voice Wakeup Driver"
+	depends on JZ_DMIC_V12
+	default n
+	help
+		Say Y here if you want to use voice wakeup function. The Voice Wakeup
+		Driver support voice wakeup. when in doubt , Say N here.
+config TEST_SECOND_REFRESH
+	bool "Ingenic Test second refresh for watch(test driver.)"
+	default n
+	help
+		This is a test driver for second refresh, this will remap tlb when system sleep.
+		do not choose this unless you know what you are doing.
+
+config DEVMEM
+	bool "Memory device driver"
+	default y
+	help
+	  The memory driver provides two character devices, mem and kmem, which
+	  provide access to the system's memory. The mem device is a view of
+	  physical memory, and each byte in the device corresponds to the
+	  matching physical address. The kmem device is the same as mem, but
+	  the addresses correspond to the kernel's virtual address space rather
+	  than physical memory. These devices are standard parts of a Linux
+	  system and most users should say Y here. You might say N if very
+	  security conscience or memory is tight.
+
 config DEVKMEM
 	bool "/dev/kmem virtual device support"
 	default y
@@ -584,6 +625,10 @@ config DEVPORT
 	depends on ISA || PCI
 	default y
 
+config DCC_TTY
+	tristate "DCC tty driver"
+	depends on ARM
+
 source "drivers/s390/char/Kconfig"
 
 config MSM_SMD_PKT
