
--- a/drivers/mtd/devices/Kconfig
+++ b/drivers/mtd/devices/Kconfig
@@ -68,6 +68,67 @@ config MTD_DATAFLASH_WRITE_VERIFY
 	  device thinks the write was successful, a bit could have been
 	  flipped accidentally due to device wear or something else.
 
+config MTD_JZ_SPI_NORFLASH
+	bool "Support ingenic spi-mtd"
+	help
+		Now support GD25LQ64C
+
+config MTD_JZ_SFC
+	tristate "Ingenic JZ series SFC driver"
+	depends on  MACH_XBURST
+	help
+		SFC driver for Ingenic JZ series SoCs
+
+config MTD_JZ_SFC_NOR
+	tristate "Ingenic JZ SFC nor"
+	depends on MTD_JZ_SFC
+	help
+		SFC driver for Ingenic JZ sfc nor
+
+config MTD_JZ_SFC_NAND
+	tristate "Ingenic JZ SFC nand"
+	depends on MTD_JZ_SFC
+	help
+		SFC driver for Ingenic JZ sfc nand
+
+config SFC_DMA
+	bool "used dma"
+	depends on MTD_JZ_SFC_NOR
+	help
+		used dma
+
+config SFC_SPEED
+    int "JZ SFC speed in MHZ"
+	range 10 200
+	default 50
+	depends on MTD_JZ_SFC
+	help
+	    Select sfc speed.
+		Examples:
+			50 =>  50 MHZ
+
+if MTD_JZ_SFC
+
+choice
+	prompt "sfc Mode"
+	help
+	  Select sfc Mode
+
+config SPI_STARDARD
+	bool "standard spi mode"
+	depends on MTD_JZ_SFC
+	help
+	  Say Y here to enable spi STANDARD MODE
+
+config SPI_QUAD
+	bool "quad spi mode"
+	depends on MTD_JZ_SFC
+	help
+	  Say Y Here to enable spi QUAD MODE
+endchoice
+
+endif
+
 config MTD_DATAFLASH_OTP
 	bool "DataFlash OTP support (Security Register)"
 	depends on MTD_DATAFLASH
