
--- a/drivers/i2c/busses/Kconfig
+++ b/drivers/i2c/busses/Kconfig
@@ -765,6 +765,135 @@ config I2C_RCAR
 	  This driver can also be built as a module.  If so, the module
 	  will be called i2c-rcar.
 
+config I2C_V12_JZ
+        tristate "Ingenic SoC based on Xburst arch's I2C controler Driver support"
+	depends on MACH_XBURST
+	default n
+	help
+	  Say Y here in order to support Ingenic SoC's I2C Controller Drvier,
+	  for example, M200, M150, JZ4775, JZ4780, and so on.
+
+config I2C_FIFO_LEN
+        int "JZ_V12 I2C Controller FIFO length"
+	default 64
+	depends on I2C_V12_JZ
+	help
+	  M200 and JZ4775 have 64 entries FIFO.
+
+config I2C_DEBUG_INFO
+	bool "enable or disable Ingenic Soc's I2C driver debug info"
+	default n
+	depends on I2C_V12_JZ
+	help
+	  I2C debug print info, you can get the debug information from
+	  /sys/devices/platform/jz_i2c.x/debug_info,
+	  for example, echo 3 > /sys/devices/platform/jz_i2c.x/debug
+
+config SOFT_I2C0_GPIO_V12_JZ
+	bool "soft gpio for i2c controler 0 Interface support"
+	depends on I2C_GPIO
+
+config I2C0_V12_JZ
+        tristate "JZ_v12 i2c controler 0 Interface support"
+	depends on I2C_V12_JZ && !SOFT_I2C0_GPIO_V12_JZ
+	default n
+
+choice
+	prompt "JZ_v12 i2c controler 0 function pins select"
+	depends on I2C0_V12_JZ
+	default I2C0_PA12_PA13
+config I2C0_PA12_PA13
+	depends on SOC_T10 || SOC_T20 || SOC_T30 || SOC_T21 || SOC_T31
+	bool "GPIO_PA12 & GPIO_PA13"
+endchoice
+
+config I2C0_NON_RESTART_MODE
+	bool "controler i2c0 no restart mode"
+	depends on I2C0_V12_JZ
+	default n
+
+config I2C0_SPEED
+        int "Jz_v12 i2c0 speed in KHZ"
+        range 100 400
+        default 400
+        depends on I2C0_V12_JZ || SOFT_I2C0_GPIO_V12_JZ
+
+config SOFT_I2C1_GPIO_V12_JZ
+	bool "soft gpio for i2c controler 1 Interface support"
+	depends on I2C_GPIO
+
+config I2C1_V12_JZ
+        tristate "JZ_v12 i2c controler 1 Interface support"
+	depends on I2C_V12_JZ && !SOFT_I2C1_GPIO_V12_JZ
+	default n
+
+choice
+	prompt "JZ_v12 i2c controler 1 function pins select"
+	depends on I2C1_V12_JZ
+	default I2C1_PB25_PB26
+config I2C1_PB25_PB26
+	depends on SOC_T10 || SOC_T20 || SOC_T30 || SOC_T21 || SOC_T31
+	bool "GPIO_PB25 & GPIO_PB26"
+config I2C1_PA16_PA17
+	depends on SOC_T31
+	bool "GPIO_PA16 & GPIO_PA17"
+config I2C1_PC08_PC09
+	depends on SOC_T31
+	bool "GPIO_PC08 & GPIO_PC09"
+endchoice
+
+config I2C1_SPEED
+        int "Jz_v12 i2c1 speed in KHZ"
+        range 10 400
+        default 400
+	depends on I2C1_V12_JZ || SOFT_I2C1_GPIO_V12_JZ
+
+config I2C1_NON_RESTART_MODE
+	bool "controler i2c1 no restart mode"
+	depends on I2C1_V12_JZ
+	default n
+
+config SOFT_I2C2_GPIO_V12_JZ
+	bool "soft gpio for i2c controler 2 Interface support"
+	depends on I2C_GPIO
+
+config I2C2_V12_JZ
+        tristate "JZ_v12 i2c controler 2 Interface support"
+	depends on I2C_V12_JZ && !SOFT_I2C2_GPIO_V12_JZ
+	default n
+
+choice
+	prompt "JZ_v12 i2c controler 2 function pins select"
+	depends on I2C2_V12_JZ
+	default I2C2_PC27_PC28
+config I2C2_PC27_PC28
+	depends on SOC_T30
+	bool "GPIO_PC27 & GPIO_PC28"
+endchoice
+
+config I2C2_SPEED
+        int "Jz_v12 i2c2 speed in KHZ"
+        range 10 400
+        default 400
+	depends on I2C2_V12_JZ || SOFT_I2C2_GPIO_V12_JZ
+
+config I2C2_NON_RESTART_MODE
+	bool "controler i2c2 no restart mode"
+	depends on I2C2_V12_JZ
+	default n
+
+config I2C_JZV12_WAIT_MS
+        int "JZ i2c bus wait timeout in ms"
+	range 10 1000
+        default 1000
+        depends on I2C0_V12_JZ || I2C1_V12_JZ
+        help
+        Select Jz bus wait time in ms.
+         Examples:
+	      100 => 100 ms
+              200 => 200 ms
+	      300 => 300 ms
+
 comment "External I2C/SMBus adapter drivers"
 
 config I2C_DIOLAN_U2C
