
--- a/drivers/power/Kconfig
+++ b/drivers/power/Kconfig
@@ -342,6 +342,25 @@ config CHARGER_SMB347
 	  Say Y to include support for Summit Microelectronics SMB347
 	  Battery Charger.
 
+config BATTERY_RICOH619
+	tristate "Ricoh R5T619 PMIC battery driver"
+	depends on MFD_RICOH619 && I2C && GENERIC_HARDIRQS
+	help
+	  Say Y to enable support for the battery control of the Ricoh R5T619
+	  Power Management device.
+choice
+	prompt "BATTERY_RICOH619 battery select"
+	depends on BATTERY_RICOH619
+	default LARGE_CAPACITY_BATTERY
+
+config LARGE_CAPACITY_BATTERY
+	  bool "Large capacity battery for 2000mA 4000mA"
+
+config SMALL_CAPACITY_BATTERY
+	  bool "Small capacity battery for 260mA 500mA"
+endchoice
+
+
 config CHARGER_TPS65090
 	tristate "TPS65090 battery charger driver"
 	depends on MFD_TPS65090
