
--- a/drivers/net/wireless/Kconfig
+++ b/drivers/net/wireless/Kconfig
@@ -7,6 +7,8 @@ menuconfig WLAN
 	depends on !S390
 	depends on NET
 	select WIRELESS
+	select WIRELESS_EXT
+	select WEXT_PRIV
 	default y
 	---help---
 	  This section contains all the pre 802.11 and 802.11 wireless
@@ -264,6 +266,11 @@ config MWL8K
 	  To compile this driver as a module, choose M here: the module
 	  will be called mwl8k.  If unsure, say N.
 
+config WIFI_CONTROL_FUNC
+	bool "Enable WiFi control function abstraction"
+	help
+	  Enables Power/Reset/Carddetect function abstraction
+
 source "drivers/net/wireless/ath/Kconfig"
 source "drivers/net/wireless/b43/Kconfig"
 source "drivers/net/wireless/b43legacy/Kconfig"
@@ -280,5 +287,9 @@ source "drivers/net/wireless/rtlwifi/Kconfig"
 source "drivers/net/wireless/ti/Kconfig"
 source "drivers/net/wireless/zd1211rw/Kconfig"
 source "drivers/net/wireless/mwifiex/Kconfig"
+source "drivers/net/wireless/bcm43341/Kconfig"
+source "drivers/net/wireless/bcmdhd_1_141_66/Kconfig"
+source "drivers/net/wireless/bcmdhd_ap6181/Kconfig"
+source "drivers/net/wireless/mt7601u/Kconfig"
 
 endif # WLAN
