
--- a/drivers/video/backlight/Kconfig
+++ b/drivers/video/backlight/Kconfig
@@ -148,6 +148,45 @@ config LCD_HX8357
 	  If you have a HX-8357 LCD panel, say Y to enable its LCD control
 	  driver.
 
+config LCD_BYD_8991FTGF
+	tristate "BYD 8991FTGF panel(480x800)"
+	depends on BACKLIGHT_CLASS_DEVICE
+	default n
+config LCD_BYD_9177AA
+	tristate "byd 9177aa  mipi-dsi panel(540x960)"
+	depends on FB && JZ_MIPI_DSI
+	help
+	   jz mipi dsi.dpi TFT display panel  support
+config LCD_X163
+   tristate "X163 smart mipi-dsi panel(320x320)"
+   depends on FB && JZ_MIPI_DSI
+   help
+	   jz mipi dsi.dbi smart display panel support
+config AUO_X163_ROTATION_180
+	bool "AUO X163 rotation 180"
+	depends on LCD_X163
+	default n
+	help
+	  If your board is needed rotation 180, select this option.
+
+choice
+        prompt "lcd type"
+        depends on SOC_T30
+        default T30_FPGA_SLCD_TRULY_240_240
+
+config T30_FPGA_SLCD_TRULY_240_240
+        bool "Truly 240*240 8bit SLCD"
+
+config T30_FPGA_SLCD_KFM701A21
+        bool "KFM701A21 400*240 16 bit SLCD"
+
+config T30_FPGA_TFT_BM8766
+        bool "BM8766 800*480 18 bit TFT"
+
+config T30_FPGA_SLCD_SPI_BYD9158B
+        bool "BYD9158B 128*160 SPI 3 or 4 bit SLCD"
+
+endchoice
 endif # LCD_CLASS_DEVICE
 
 #
@@ -261,10 +300,10 @@ config BACKLIGHT_CARILLO_RANCH
 
 config BACKLIGHT_PWM
 	tristate "Generic PWM based Backlight Driver"
-	depends on PWM
+	depends on PWM && JZ_PWM
 	help
 	  If you have a LCD backlight adjustable by PWM, say Y to enable
-	  this driver.
+	  this driver. Choice JZ PWM driver and the PWM channel.
 
 config BACKLIGHT_DA903X
 	tristate "Backlight Driver for DA9030/DA9034 using WLED"
