
--- a/drivers/watchdog/Kconfig
+++ b/drivers/watchdog/Kconfig
@@ -1028,6 +1028,13 @@ config JZ4740_WDT
 	help
 	  Hardware driver for the built-in watchdog timer on Ingenic jz4740 SoCs.
 
+config JZ_WDT
+	tristate "Ingenic jz SoC hardware watchdog"
+	depends on MACH_XBURST
+	select WATCHDOG_CORE
+	help
+	  Hardware driver for the built-in watchdog timer on Ingenic jz SoCs.
+
 config WDT_MTX1
 	tristate "MTX-1 Hardware Watchdog"
 	depends on MIPS_MTX1
