
--- a/drivers/mmc/host/Kconfig
+++ b/drivers/mmc/host/Kconfig
@@ -4,6 +4,357 @@
 
 comment "MMC/SD/SDIO Host Controller Drivers"
 
+config JZMMC_V12
+	tristate "Ingenic(XBurst)  MMC/SD Card Controller(MSC) v1.2 support"
+	depends on SOC_M200 || SOC_T15 || SOC_T10 || SOC_T20 || SOC_T30 || SOC_T21 || SOC_T31
+	help
+	  This selects the Ingenic XBurst SD/MMC Card Controller v1.2.
+	  If you have platform with a SD/Multimedia Card slot and compact
+	  with this version, say Y or M here.
+	  If unsure, say N.
+choice
+	prompt "JZMMC_V12 data transfer method select"
+	depends on JZMMC_V12
+	default JZMMC_V12_SDMA
+config JZMMC_V12_SDMA
+	bool "JZMMC_V12 data transfer using MSC SDMA"
+endchoice
+
+config JZMMC_V12_MMC0
+	bool "JZMMC_V12 MMC0"
+	depends on JZMMC_V12
+choice
+	prompt "JZMMC_V12 GPIO function pins select"
+	depends on JZMMC_V12_MMC0
+	default JZMMC_V12_MMC0_PA_4BIT
+config JZMMC_V12_MMC0_PA_4BIT
+	bool "GPIO A, data with 4 bit"
+config JZMMC_V12_MMC0_PA_8BIT
+	bool "GPIO A, data with 8 bit"
+config JZMMC_V12_MMC0_PB_4BIT
+	bool "GPIO B, data with 4 bit"
+endchoice
+
+config MMC0_MAX_FREQ
+	int "msc0 max frequency"
+	depends on JZMMC_V12_MMC0
+	default 24000000
+	help
+	  The max frequency of msc0, it should NOT be higher than 50mHz.
+	  If unsure, say N.
+
+config MMC0_PIO_MODE
+	bool "Use PIO mode for MSC0"
+	depends on JZMMC_V12_MMC0 && MMC_BLOCK_BOUNCE
+	default n
+	help
+	  Use PIO mode to transfer data for MSC0, say Y or N here.
+	  If unsure, say N.
+
+config JZMMC_V12_MMC1
+	bool "JZMMC_V12 MMC1"
+	depends on JZMMC_V12
+choice
+	depends on JZMMC_V12_MMC1
+	prompt "JZMMC_V12 GPIO function pins select"
+	default JZMMC_V12_MMC1_PB_4BIT
+config JZMMC_V12_MMC1_PA_4BIT
+	bool "GPIO A, Data width 4 bit"
+config JZMMC_V12_MMC1_PD_4BIT
+	bool "GPIO D, Data width 4 bit"
+config JZMMC_V12_MMC1_PB_4BIT
+	bool "GPIO B, Data width 4 bit"
+config JZMMC_V12_MMC1_PC_4BIT
+	bool "GPIO C, Data width 4 bit"
+endchoice
+
+config MMC1_MAX_FREQ
+	int "MSC1 max frequency"
+	depends on JZMMC_V12_MMC1
+	default 24000000
+	help
+	  The max frequency of msc1, it should not be higher than 50 Mhz
+	  If unsure, say N.
+
+config MMC1_PIO_MODE
+	bool "Use PIO mode for MSC1"
+	depends on JZMMC_V12_MMC1 && MMC_BLOCK_BOUNCE
+	default n
+	help
+	 Use PIO mode to transfer data for MSC1, say Y or N here.
+	 If unsure, say N.
+
+config JZMMC_V12_MMC2
+	bool "JZMMC_V12 MMC2"
+	depends on JZMMC_V12
+choice
+	prompt "JZMMC_V12 GPIO function pins select"
+	depends on JZMMC_V12_MMC2
+	default JZMMC_V12_MMC2_PE_4BIT
+config JZMMC_V12_MMC2_PB_4BIT
+	bool "GPIO B, Data width 4 bit"
+config JZMMC_V12_MMC2_PE_4BIT
+	bool "GPIO E, Data width 4 bit"
+endchoice
+
+config MMC2_MAX_FREQ
+	int "MSC2 max frequency"
+	depends on JZMMC_V12_MMC2
+	default 24000000
+	help
+	 The max frequency of msc2, it should not be higher than 50 Mhz,
+	 If unsure , say N.
+
+config MMC2_PIO_MODE
+	bool "Use PIO mode for MSC2"
+	depends on JZMMC_V12_MMC2 && MMC_BLOCK_BOUNCE
+	default n
+	help
+	 Use PIO mode to transfer data for MSC2, say Y or N here.
+	 If unsure, say N.
+
+config MMC_JZ4780
+	tristate "Ingenic(XBurst) JZ4780 MMC/SD Card Controller(MSC) support"
+	depends on SOC_4780
+	help
+	  This selects the Ingenic XBurst SD/MMC Card Controller
+	  for JZ4780 support. If you have an JZ4780
+	  platform with a SD/Multimedia Card slot, say Y or M here.
+
+	  If unsure, say N.
+
+config MMC0_JZ4780
+	bool "MSC0"
+	depends on MMC_JZ4780
+	help
+	  If you want enable MSC0, say Y or M here.
+
+	  If unsure, say N.
+choice
+	prompt "JZ4780 MSC0 function pins select"
+	depends on MMC0_JZ4780
+	default MMC0_JZ4780_PA_4BIT
+config MMC0_JZ4780_PA_4BIT
+	bool "GPIO A, Data width 4 bit"
+config MMC0_JZ4780_PA_8BIT
+	bool "GPIO A, Data width 8 bit"
+config MMC0_JZ4780_PE_4BIT
+	bool "GPIO E, Data width 4 bit"
+config MMC0_JZ4780_PA_4BIT_RESET
+	bool "GPIO A, Data width 4 bit, use Reset pin(eMMC)"
+config MMC0_JZ4780_PA_8BIT_RESET
+	bool "GPIO A, Data width 8 bit, use Reset pin(eMMC)"
+endchoice
+
+config MMC0_MAX_FREQ
+	int "MSC0 max frequency"
+	depends on MMC0_JZ4780
+	default 24000000
+	help
+	  The max frequency of msc0, it should NOT be higher than 50mHz.
+
+	  If unsure, say N.
+
+config MMC0_PIO_MODE
+	bool "Use PIO mode for MSC0"
+	depends on MMC0_JZ4780 && MMC_BLOCK_BOUNCE
+	default n
+	help
+	  Use PIO mode to transfer data for MSC0, say Y or N here.
+
+	  If unsure, say N.
+
+config MMC1_JZ4780
+	bool "MSC1"
+	depends on MMC_JZ4780
+	help
+	  If you want enable MSC1, say Y or M here.
+
+	  If unsure, say N.
+choice
+	prompt "JZ4780 MSC1 function pins select"
+	depends on MMC1_JZ4780
+	default MMC1_JZ4780_PD_4BIT
+config MMC1_JZ4780_PD_4BIT
+	bool "GPIO D, Data width 4 bit"
+config MMC1_JZ4780_PE_4BIT
+	bool "GPIO E, Data width 4 bit"
+endchoice
+
+config MMC1_MAX_FREQ
+	int "MSC1 max frequency"
+	depends on MMC1_JZ4780
+	default 24000000
+	help
+	  The max frequency of msc1, it should NOT be higher than 50mHz.
+
+	  If unsure, say N.
+
+config MMC1_PIO_MODE
+	bool "Use PIO mode for MSC1"
+	depends on MMC1_JZ4780 && MMC_BLOCK_BOUNCE
+	default n
+	help
+	  Use PIO mode to transfer data for MSC1, say Y or N here.
+
+	  If unsure, say N.
+
+config MMC2_JZ4780
+	bool "MSC2"
+	depends on MMC_JZ4780
+	help
+	  If you want enable MSC2, say Y or M here.
+
+	  If unsure, say N.
+choice
+	prompt "JZ4780 MSC2 function pins select"
+	depends on MMC2_JZ4780
+	default MMC2_JZ4780_PB_4BIT
+config MMC2_JZ4780_PB_4BIT
+	bool "GPIO B, Data width 4 bit"
+config MMC2_JZ4780_PE_4BIT
+	bool "GPIO E, Data width 4 bit"
+endchoice
+
+config MMC2_MAX_FREQ
+	int "MSC2 max frequency"
+	depends on MMC2_JZ4780
+	default 24000000
+	help
+	  The max frequency of msc2, it should NOT be higher than 50mHz.
+
+	  If unsure, say N.
+
+config MMC2_PIO_MODE
+	bool "Use PIO mode for MSC2"
+	depends on MMC2_JZ4780 && MMC_BLOCK_BOUNCE
+	default n
+	help
+	  Use PIO mode to transfer data for MSC2, say Y or N here.
+
+	  If unsure, say N.
+
+
+config JZMMC_V11
+	tristate "Ingenic(XBurst) Halley MMC/SD Card Controller(MSC) support"
+	depends on SOC_4775
+	help
+	  This selects the Ingenic XBurst SD/MMC Card Controller
+	  for Halley support. If you have an Halley
+	  platform with a SD/Multimedia Card slot, say Y or M here.
+
+	  If unsure, say N.
+
+config JZMMC_V11_MMC0
+	bool "MSC0"
+	depends on JZMMC_V11
+	help
+	  If you want enable MSC0, say Y or M here.
+
+	  If unsure, say N.
+choice
+	prompt "Halley MSC0 function pins select"
+	depends on JZMMC_V11_MMC0
+	default JZMMC_V11_MMC0_PA_4BIT
+config JZMMC_V11_MMC0_PA_4BIT
+	bool "GPIO A, Data width 4 bit"
+config JZMMC_V11_MMC0_PA_8BIT
+	bool "GPIO A, Data width 8 bit"
+config JZMMC_V11_MMC0_PE_4BIT
+	bool "GPIO E, Data width 4 bit"
+endchoice
+
+config MMC0_MAX_FREQ
+	int "MSC0 max frequency"
+	depends on JZMMC_V11_MMC0
+	default 24000000
+	help
+	  The max frequency of msc0, it should NOT be higher than 50mHz.
+
+	  If unsure, say N.
+
+config MMC0_PIO_MODE
+	bool "Use PIO mode for MSC0"
+	depends on JZMMC_V11_MMC0 && MMC_BLOCK_BOUNCE
+	default n
+	help
+	  Use PIO mode to transfer data for MSC0, say Y or N here.
+
+	  If unsure, say N.
+
+config JZMMC_V11_MMC1
+	bool "MSC1"
+	depends on JZMMC_V11
+	help
+	  If you want enable MSC1, say Y or M here.
+
+	  If unsure, say N.
+choice
+	prompt "Halley MSC1 function pins select"
+	depends on JZMMC_V11_MMC1
+	default JZMMC_V11_MMC1_PD_4BIT
+config JZMMC_V11_MMC1_PD_4BIT
+	bool "GPIO D, Data width 4 bit"
+config JZMMC_V11_MMC1_PE_4BIT
+	bool "GPIO E, Data width 4 bit"
+endchoice
+
+config MMC1_MAX_FREQ
+	int "MSC1 max frequency"
+	depends on JZMMC_V11_MMC1
+	default 24000000
+	help
+	  The max frequency of msc1, it should NOT be higher than 50mHz.
+
+	  If unsure, say N.
+
+config MMC1_PIO_MODE
+	bool "Use PIO mode for MSC1"
+	depends on JZMMC_V11_MMC1 && MMC_BLOCK_BOUNCE
+	default n
+	help
+	  Use PIO mode to transfer data for MSC1, say Y or N here.
+
+	  If unsure, say N.
+
+config JZMMC_V11_MMC2
+	bool "MSC2"
+	depends on JZMMC_V11
+	help
+	  If you want enable MSC2, say Y or M here.
+
+	  If unsure, say N.
+choice
+	prompt "Halley MSC2 function pins select"
+	depends on JZMMC_V11_MMC2
+	default JZMMC_V11_MMC2_PB_4BIT
+config JZMMC_V11_MMC2_PB_4BIT
+	bool "GPIO B, Data width 4 bit"
+config JZMMC_V11_MMC2_PE_4BIT
+	bool "GPIO E, Data width 4 bit"
+endchoice
+
+config MMC2_MAX_FREQ
+	int "MSC2 max frequency"
+	depends on JZMMC_V11_MMC2
+	default 24000000
+	help
+	  The max frequency of msc2, it should NOT be higher than 50mHz.
+
+	  If unsure, say N.
+
+config MMC2_PIO_MODE
+	bool "Use PIO mode for MSC2"
+	depends on JZMMC_V11_MMC2 && MMC_BLOCK_BOUNCE
+	default n
+	help
+	  Use PIO mode to transfer data for MSC2, say Y or N here.
+
+	  If unsure, say N.
+
+
+
+
 config MMC_ARMMMCI
 	tristate "ARM AMBA Multimedia Card Interface support"
 	depends on ARM_AMBA
