
--- /dev/null
+++ b/drivers/video/avpu/jz_avpu/Kconfig
@@ -0,0 +1,11 @@
+    config AVPU_DRIVER
+        bool "AVPU driver"
+        depends on HAS_DMA
+        depends on HAS_IOMEM
+		default n
+        help
+	  Say Y for enabling the hardware codec driver
+
+          To compile this support as a module, choose M here:
+          The module will be called t31_encoder.
+          If unsure, say N.
