
--- /dev/null
+++ b/drivers/net/wireless/mt7601u/PREALLOC/os/linux/Makefile.clean
@@ -0,0 +1,20 @@
+include $(RT28xx_DIR)/os/linux/config.mk
+
+PHONY := clean install uninstall
+
+clean:
+	rm -f ../../prealloc.o
+	rm -f ../../common/.*.cmd .*.flags .*.d
+	rm -f ../../os/linux/*.o *.ko *.mod.o *.mod.c
+	rm -f ../../os/linux/.*.cmd .*.flags .*.d
+	rm -fr ../../os/linux/.tmp_versions
+#Must clean Module.symvers; or you will suffer symbol version not match
+#when OS_ABL = YES.
+	rm -f ../../os/linux/Module.symvers
+	rm -f ../../os/linux/Modules.symvers
+	rm -f ../../os/linux/Module.markers
+	rm -f ../../os/linux/modules.order
+
+# Declare the contents of the .PHONY variable as phony.  We keep that
+# information in a variable so we can use it in if_changed and friends.
+.PHONY: $(PHONY)
