
--- a/drivers/misc/Makefile
+++ b/drivers/misc/Makefile
@@ -36,8 +36,10 @@ obj-$(CONFIG_SENSORS_TSL2550)	+= tsl2550.o
 obj-$(CONFIG_EP93XX_PWM)	+= ep93xx_pwm.o
 obj-$(CONFIG_DS1682)		+= ds1682.o
 obj-$(CONFIG_TI_DAC7512)	+= ti_dac7512.o
+obj-$(CONFIG_UID_STAT)		+= uid_stat.o
 obj-$(CONFIG_C2PORT)		+= c2port/
 obj-$(CONFIG_HMC6352)		+= hmc6352.o
+obj-$(CONFIG_JZ_SPI_NOR)        += jz_spi_nor/
 obj-y				+= eeprom/
 obj-y				+= cb710/
 obj-$(CONFIG_SPEAR13XX_PCIE_GADGET)	+= spear13xx_pcie_gadget.o
@@ -47,9 +49,20 @@ obj-$(CONFIG_PCH_PHUB)		+= pch_phub.o
 obj-y				+= ti-st/
 obj-y				+= lis3lv02d/
 obj-y				+= carma/
+obj-$(CONFIG_BCM_PM_CORE) 	+= bcm-pm-core.o
+obj-$(CONFIG_BROADCOM_RFKILL)   += bt-power-bluesleep.o
+obj-$(CONFIG_GPS_RFKILL)    += gps-power.o
 obj-$(CONFIG_USB_SWITCH_FSA9480) += fsa9480.o
 obj-$(CONFIG_ALTERA_STAPL)	+=altera-stapl/
 obj-$(CONFIG_INTEL_MEI)		+= mei/
 obj-$(CONFIG_VMWARE_VMCI)	+= vmw_vmci/
+obj-$(CONFIG_JZ_EFUSE_V13)  += jz_efuse_v13.o
+obj-$(CONFIG_JZ_IRDA_V11)   += jz_irda_v11.o
 obj-$(CONFIG_LATTICE_ECP3_CONFIG)	+= lattice-ecp3-config.o
 obj-$(CONFIG_SRAM)		+= sram.o
+obj-$(CONFIG_LOGGER)		+= logger.o
+obj-$(CONFIG_LINUX_PMEM)	+= linux_pmem.o
+obj-$(CONFIG_JZ_AES)		+= jz_aes/
+obj-$(CONFIG_JZ_DES)		+= jz_des/
+obj-$(CONFIG_RMEM)		+= rmem.o
+obj-$(CONFIG_PWM_SDK)		+= pwm_sdk.o
