
--- /dev/null
+++ b/drivers/video/jz_fb_v12/Kconfig
@@ -0,0 +1,81 @@
+# JZ LCD V1.2  configuration
+
+menuconfig FB_JZ_V12
+	tristate "JZ LCDC framebuffer V1.2"
+	depends on FB
+	select FB_CFB_FILLRECT
+	select FB_CFB_COPYAREA
+	select FB_CFB_IMAGEBLIT
+	help
+	  Framebuffer support for the M200 SoC.
+
+config JZ_MIPI_DSI
+	bool "jz mipi dsi interface"
+    depends on FB_JZ_V12
+	default n
+
+
+config JZ_MIPI_DBI
+	bool "jz mipi dbi support mipi slcd"
+    depends on JZ_MIPI_DSI
+	default n
+
+config DEFAULT_DATALANE_BPS
+	int "set default datalane Mbps"
+	range 90 1000
+	depends on JZ_MIPI_DSI
+	default 432
+
+config DSI_DPI_DEBUG
+	bool "only for tft.dpi interface debug and color test pattern"
+	depends on JZ_MIPI_DSI
+	default n
+
+choice
+	depends on FB_JZ_V12 && (!JZ_MIPI_DSI)
+	prompt "set lcd gpio"
+	default LCD_GPIO_FUNC0_24BIT
+
+config LCD_GPIO_FUNC0_16BIT
+	bool "func0 with 16bit bpp"
+	help
+	 set lcd gpio function0 with 16bit bpp
+		for example,
+		for BYD_BM8766U panel, 16bpp, set it
+
+config LCD_GPIO_FUNC0_18BIT
+	bool "func0 with 18bit bpp"
+	help
+	 set lcd gpio function0 with 18bit bpp
+		for example,
+		for BM347WV_F_8991FTGF_HX8369 panel, you can set it
+
+config LCD_GPIO_FUNC0_24BIT
+	bool "func0 with 24bit bpp"
+	help
+	 set lcd gpio function0 with 24bit bpp
+		for example,
+		for BYD_BM8766U panel, 24bpp, set it
+
+config LCD_GPIO_FUNC2_SLCD
+	bool "func2 only for 16bit slcd"
+	help
+	 set lcd gpio function2 only for 16bit smart panel
+		for example,
+		for KFM7021a21 panel, 16bpp, set it
+
+config LCD_GPIO_FUNC2_SLCD_8BIT
+	bool "func2 only for 8bit slcd"
+	help
+	 set lcd gpio function2 only for 8bit smart panel
+		for example,
+		for truly_tft240240_2_e panel, set it
+endchoice
+
+config FB_JZ_DEBUG
+        bool "fb test for displaying color bar"
+        depends on FB_JZ_V12
+        default n
+        help
+           fb test for displaying color bar in your board.
+
