
--- a/drivers/base/power/wakeup.c
+++ b/drivers/base/power/wakeup.c
@@ -659,6 +659,23 @@ void pm_wakeup_event(struct device *dev, unsigned int msec)
 }
 EXPORT_SYMBOL_GPL(pm_wakeup_event);
 
+int get_active_wakeup_sources(char *buf)
+{
+	struct wakeup_source *ws;
+	int count = 0;
+
+	count += sprintf(buf+count,"active wakeup source:\n");
+	rcu_read_lock();
+	list_for_each_entry_rcu(ws, &wakeup_sources, entry) {
+		if (ws->active) {
+			count += sprintf(buf+count, "%s ", ws->name);
+		}
+	}
+	rcu_read_unlock();
+	count += sprintf(buf+count,"\n");
+	return count;
+}
+
 static void print_active_wakeup_sources(void)
 {
 	struct wakeup_source *ws;
