
--- a/drivers/input/touchscreen/Kconfig
+++ b/drivers/input/touchscreen/Kconfig
@@ -10,6 +10,97 @@ menuconfig INPUT_TOUCHSCREEN
 	  If unsure, say Y.
 
 if INPUT_TOUCHSCREEN
+############-ingenic-touchscreen-############
+config M200_SUPPORT_TSC
+	bool "Touchscreen Supported on M200"
+	depends on SOC_M200 && INPUT_TOUCHSCREEN
+	help
+	  This selects the touchscreen that Ingenic XBurst M200 support,
+	  say Y or N here.
+
+	  If unsure, say N.
+
+config TOUCHSCREEN_FT5336
+	tristate "FocalTech FT5336 TouchScreen driver"
+	depends on I2C && M200_SUPPORT_TSC && I2C0_V12_JZ
+	default n
+	help
+	  FocalTech FT5336GQQ TouchSreen Driver implemented by Ingenic.
+config TOUCHSCREEN_FT6X06
+	tristate "FocalTech FT6X06 TouchScreen driver"
+	depends on I2C && M200_SUPPORT_TSC && I2C0_V12_JZ
+	default n
+	help
+	  pisecs FocalTech FT6X06 TouchSreen Driver implemented by Ingenic.
+
+config FT6X06_MULTITOUCH
+	bool "Enable multi-touch on FocalTech FT6X06 "
+	depends on TOUCHSCREEN_FT6X06
+	default y
+	help
+	  support multitouch ft6206,if you want , say Y here.
+
+config TOUCHSCREEN_FT6X0X
+       tristate "Touchscreen ft6x0x"
+       depends on I2C && M200_SUPPORT_TSC
+       help
+       ft6x0x is a CTP controller
+
+if TOUCHSCREEN_FT6X0X
+config FT6X0X_EXT_FUNC
+	 bool "ft6x0x extra function"
+	 depends on TOUCHSCREEN_FT6X0X
+	 default n
+	 help
+		ft6x0x extra function:
+		l.sysfs inferface
+		2.tp firmware upgrade control
+		3.tp regs read write
+
+if FT6X0X_EXT_FUNC
+choice
+	 prompt "ft6x0x fireware"
+		  depends on FT6X0X_EXT_FUNC
+		  default TOUCHSCREEN_FT6X0X_YIYANG_1S
+config TOUCHSCREEN_FT6X0X_YIYANG_1S
+	 bool "YIYANG fireware monitor 1S"
+
+config TOUCHSCREEN_FT6X0X_YIYANG_10S
+	 bool "YIYANG fireware monitor 10S"
+
+config TOUCHSCREEN_FT6X0X_TWS
+	 bool "TWS fireware"
+endchoice
+endif
+endif
+
+config 	TSC_SWAP_XY
+		bool "swap x and y "
+		depends on INPUT_TOUCHSCREEN
+		default n
+		help
+			say Y here if you need flip x y.
+config 	TSC_SWAP_X
+		bool "swap x"
+		depends on INPUT_TOUCHSCREEN
+		default n
+		help
+			say Y here if you need flip x y.
+config 	TSC_SWAP_Y
+		bool "swap y "
+		depends on INPUT_TOUCHSCREEN
+		default n
+		help
+			say Y here if you need flip x y.
+
+config FT5X0X_DEBUG
+		bool "Enable FT5X0X TouchScreen driver Debug output"
+		depends on TOUCHSCREEN_FT5306 || TOUCHSCREEN_FT5X0X
+		default n
+		help
+			say Y here if you are trying to debug cft5306
+
+#############################################
 
 config TOUCHSCREEN_88PM860X
 	tristate "Marvell 88PM860x touchscreen"
@@ -66,6 +157,19 @@ config TOUCHSCREEN_AD7879
 	  To compile this driver as a module, choose M here: the
 	  module will be called ad7879.
 
+config TOUCHSCREEN_ITE7258
+	tristate "Touchscreen ite7258"
+	depends on I2C && M200_SUPPORT_TSC
+	help
+		ite7258 a CTP controller,support panel unknown
+
+config ITE7258_MULTITOUCH
+	bool "Enable multi-touch on FocalTech ITE7258 "
+	depends on TOUCHSCREEN_ITE7258
+	default y
+	help
+	  support multitouch ite7258,if you want , say Y here.
+
 config TOUCHSCREEN_AD7879_I2C
 	tristate "support I2C bus connection"
 	depends on TOUCHSCREEN_AD7879 && I2C
@@ -86,6 +190,7 @@ config TOUCHSCREEN_AD7879_SPI
 	  To compile this driver as a module, choose M here: the
 	  module will be called ad7879-spi.
 
+
 config TOUCHSCREEN_ATMEL_MXT
 	tristate "Atmel mXT I2C Touchscreen"
 	depends on I2C
