
--- /dev/null
+++ b/drivers/usb/dwc2/Kconfig
@@ -0,0 +1,98 @@
+config USB_JZ_DWC2
+	tristate "DesignWare USB2 DRD Core Support"
+	depends on USB_SUPPORT
+	help
+	  Say Y or M here if your system has a Dual Role SuperSpeed
+	  USB controller based on the DesignWare USB2.0 IP Core.
+
+	  If you choose to build this driver is a dynamically linked
+	  module, the module will be called dwc2.ko.
+
+if USB_JZ_DWC2
+
+choice
+	prompt "Driver Mode"
+	help
+	  Select Driver Mode
+
+config USB_DWC2_DUAL_ROLE
+	bool "Both Host and Device"
+	depends on (USB && USB_GADGET)
+	select USB_OTG
+	help
+	  Say Y here to enable both Host and Device function
+
+config USB_DWC2_HOST_ONLY
+	bool "Host Mode Only"
+	depends on USB
+	help
+	  Say Y here to enable Host function only
+
+config USB_DWC2_DEVICE_ONLY
+	bool "Device Mode Only"
+	depends on USB_GADGET
+	help
+	  Say Y Here to enable Device function only
+
+endchoice
+
+config BOARD_HAS_NO_DETE_FACILITY
+	bool "Board has no plug detect facility"
+	default n
+	help
+		Say Y here if your board has no plug detect facility
+
+config CONFIG_DWC2_HOST_OVERCURRENT_DET
+	bool "hcd overcurrent detect (vbus)"
+	default n
+	help
+		Say Y here if your board have vbus input and you want hcd overcurrent detect
+
+config USB_DWC2_DEBUG
+	bool "Enable Debugging(dev_dbg) Messages"
+	help
+		Say Y here to enable debugging messages on DWC2 Driver.
+
+config USB_DWC2_VERBOSE
+	bool "Enable Verbose Debugging(dev_vdbg) Messages"
+	depends on USB_DWC2_DEBUG
+	help
+	  Say Y here to enable verbose debugging messages on DWC2 Driver.
+
+config USB_CHARGER_SOFTWARE_JUDGE
+	bool "Enable software to judge connect to usb adapter or host"
+	depends on (USB_DWC2_DUAL_ROLE || USB_DWC2_DEVICE_ONLY)
+	default n
+	help
+		Say Y here if you want increase usb charger current limit,
+		when connect to adapter
+
+config USB_DWC2_VERBOSE_VERBOSE
+	bool "Enable Very Verbose Debugging Messagers"
+	default y
+	help
+	  enable debugging messages for developer debug, these debugging messages
+          are switchable write 1 to file under /sys/module/dwc2/parameters/,
+          the default value is 0, so you can always say Y here and won't get boring.
+
+config USB_DWC2_HOST_TRACER
+	bool "Enable tracing host operation flow"
+	depends on (USB_DWC2_DUAL_ROLE || USB_DWC2_HOST_ONLY)
+	help
+	  Say Y here to enable tracing host operation flow,
+	  you must reserve 192M~256M memory
+
+config USB_DWC2_FULLSPEED_DEVICE
+	bool "dwc2 full speed gadget"
+	depends on (USB_DWC2_DUAL_ROLE || USB_DWC2_DEVICE_ONLY)
+	default n
+	help
+		Say Y here to keep the gadget at full speed
+
+config USB_DWC2_FULLSPEED_HOST
+	bool "dwc2 full speed host"
+	depends on (USB_DWC2_DUAL_ROLE || USB_DWC2_HOST_ONLY)
+	default n
+	help
+		Say Y here to keep the host at full speed (ohci mode)
+endif
