
--- /dev/null
+++ b/drivers/net/wireless/bcmdhd_1_141_66/Kconfig
@@ -0,0 +1,65 @@
+config BCMDHD_1_141_66
+	tristate "Broadcom bcm43438 wireless cards support"
+	---help---
+	  This module adds support for wireless adapters based on
+	  Broadcom bcm43438 chipset.
+
+	  If you choose to build a module, it'll be called dhd. Say M if
+	  unsure.
+
+config BCMDHD_1_141_66_SDIO
+	bool "SDIO bus interface support"
+	depends on BCMDHD_1_141_66 && MMC
+	default y
+
+config BCMDHD_1_141_66_PCIE
+	bool "PCIe bus interface support"
+	depends on BCMDHD_1_141_66 && PCI && !BCMDHD_1_141_66_SDIO
+
+config BCM4339
+	tristate "BCM4339 support"
+	depends on BCMDHD_1_141_66
+
+config BCM4343
+	tristate "BCM4343 support"
+	depends on BCMDHD_1_141_66
+
+config BCM4354
+	tristate "BCM4354 support"
+	depends on BCMDHD_1_141_66
+
+config BCMDHD_1_141_66_FW_PATH
+	depends on BCMDHD_1_141_66
+	string "Firmware path"
+	default "/lib/firmware/fw_bcm43438.bin"
+	---help---
+	  Path to the firmware file.
+
+config BCMDHD_1_141_66_NVRAM_PATH
+	depends on BCMDHD_1_141_66
+	string "NVRAM path"
+	default "/lib/firmware/bcm43438.cal"
+	---help---
+	  Path to the calibration file.
+
+config BCMDHD_1_141_66_WEXT
+	bool "Enable WEXT support"
+	depends on BCMDHD_1_141_66 && CFG80211 = n
+	select WIRELESS_EXT
+	select WEXT_PRIV
+	help
+	  Enables WEXT support
+
+config DHD_USE_STATIC_BUF
+	bool "Enable memory preallocation"
+	depends on BCMDHD_1_141_66
+	default n
+	---help---
+	  Use memory preallocated in platform
+
+config DHD_USE_SCHED_SCAN
+	bool "Use CFG80211 sched scan"
+	depends on BCMDHD_1_141_66 && CFG80211
+	default n
+	---help---
+	  Use CFG80211 sched scan
