
--- a/drivers/i2c/busses/Makefile
+++ b/drivers/i2c/busses/Makefile
@@ -76,6 +76,13 @@ obj-$(CONFIG_I2C_XILINX)	+= i2c-xiic.o
 obj-$(CONFIG_I2C_XLR)		+= i2c-xlr.o
 obj-$(CONFIG_I2C_RCAR)		+= i2c-rcar.o
 
+obj-$(CONFIG_I2C0_V12_JZ) 	+= i2c-v12-jz.o
+obj-$(CONFIG_I2C1_V12_JZ)       += i2c-v12-jz.o
+obj-$(CONFIG_I2C2_V12_JZ)       += i2c-v12-jz.o
+obj-$(CONFIG_I2C3_V12_JZ)       += i2c-v12-jz.o
+
+
+
 # External I2C/SMBus adapter drivers
 obj-$(CONFIG_I2C_DIOLAN_U2C)	+= i2c-diolan-u2c.o
 obj-$(CONFIG_I2C_PARPORT)	+= i2c-parport.o
