
--- a/drivers/net/phy/smsc.c
+++ b/drivers/net/phy/smsc.c
@@ -71,12 +71,13 @@ static int smsc_phy_config_init(struct phy_device *phydev)
 	if (rc < 0)
 		return rc;
 
+#if 0
 	/* Enable energy detect mode for this SMSC Transceivers */
 	rc = phy_write(phydev, MII_LAN83C185_CTRL_STATUS,
 		       rc | MII_LAN83C185_EDPWRDOWN);
 	if (rc < 0)
 		return rc;
-
+#endif
 	return smsc_phy_ack_interrupt (phydev);
 }
 
@@ -105,12 +106,12 @@ static int lan87xx_read_status(struct phy_device *phydev)
 		int rc = phy_read(phydev, MII_LAN83C185_CTRL_STATUS);
 		if (rc < 0)
 			return rc;
-
+#if 0
 		rc = phy_write(phydev, MII_LAN83C185_CTRL_STATUS,
 			       rc & ~MII_LAN83C185_EDPWRDOWN);
 		if (rc < 0)
 			return rc;
-
+#endif
 		/* Sleep 64 ms to allow ~5 link test pulses to be sent */
 		msleep(64);
 
@@ -118,11 +119,12 @@ static int lan87xx_read_status(struct phy_device *phydev)
 		rc = phy_read(phydev, MII_LAN83C185_CTRL_STATUS);
 		if (rc < 0)
 			return rc;
-
+#if 0
 		rc = phy_write(phydev, MII_LAN83C185_CTRL_STATUS,
 			       rc | MII_LAN83C185_EDPWRDOWN);
 		if (rc < 0)
 			return rc;
+#endif
 	}
 
 	return err;
