
--- a/drivers/media/platform/Kconfig
+++ b/drivers/media/platform/Kconfig
@@ -53,6 +53,18 @@ config VIDEO_VIU
 	  Say Y here if you want to enable VIU device on MPC5121e Rev2+.
 	  In doubt, say N.
 
+config VIDEO_TX_ISP
+	  bool "TX ISP MODULE"
+	  select VIDEOBUF2_CORE
+	  select VIDEOBUF2_MEMOPS
+	  depends on VIDEO_V4L2 && I2C && VIDEO_V4L2_SUBDEV_API
+	  ---help---
+	  Driver for TX ISP.
+
+if VIDEO_TX_ISP
+source "drivers/media/platform/tx-isp/Kconfig"
+endif #VIDEO_TX_ISP
+
 config VIDEO_TIMBERDALE
 	tristate "Support for timberdale Video In/LogiWIN"
 	depends on VIDEO_V4L2 && I2C && DMADEVICES
