
--- /dev/null
+++ b/drivers/net/wireless/mt7601u/common/rtusb_dev_id.c
@@ -0,0 +1,48 @@
+/****************************************************************************
+ * Ralink Tech Inc.
+ * 4F, No. 2 Technology 5th Rd.
+ * Science-based Industrial Park
+ * Hsin-chu, Taiwan, R.O.C.
+ * (c) Copyright 2002, Ralink Technology, Inc.
+ *
+ * All rights reserved. Ralink's source code is an unpublished work and the
+ * use of a copyright notice does not imply otherwise. This source code
+ * contains confidential trade secret material of Ralink Tech. Any attemp
+ * or participation in deciphering, decoding, reverse engineering or in any
+ * way altering the source code is stricitly prohibited, unless the prior
+ * written consent of Ralink Technology, Inc. is obtained.
+ ****************************************************************************
+
+    Module Name:
+    rtusb_dev_id.c
+
+    Abstract:
+
+    Revision History:
+    Who        When          What
+    ---------  ----------    ----------------------------------------------
+ */
+
+#define RTMP_MODULE_OS
+
+/*#include "rt_config.h"*/
+#include "rtmp_comm.h"
+#include "rt_os_util.h"
+#include "rt_os_net.h"
+
+
+/* module table */
+USB_DEVICE_ID rtusb_dev_id[] = {
+#ifdef RT6570
+	{USB_DEVICE(0x148f,0x6570)}, /* Ralink 6570 */
+#endif /* RT6570 */
+	{USB_DEVICE(0x148f, 0x7650)}, /* MT7650 */
+#ifdef MT7601U
+	{USB_DEVICE(0x148f,0x6370)}, /* Ralink 6370 */
+	{USB_DEVICE(0x148f,0x7601)}, /* MT 6370 */
+#endif /* MT7601U */
+	{ }/* Terminating entry */
+};
+
+INT const rtusb_usb_id_len = sizeof(rtusb_dev_id) / sizeof(USB_DEVICE_ID);
+MODULE_DEVICE_TABLE(usb, rtusb_dev_id);
