
--- a/include/linux/clk.h
+++ b/include/linux/clk.h
@@ -170,6 +170,14 @@ struct clk *devm_clk_get(struct device *dev, const char *id);
  */
 int clk_enable(struct clk *clk);
 
+/**
+ * clk_is_enable - indicate that whether the clock source is enabled.
+ * @clk: clock source
+ *
+ * Returns 1 if clock source is enabled, or 0 if disabled.
+ */
+int clk_is_enabled(struct clk *clk);
+
 /**
  * clk_disable - inform the system when the clock source is no longer required.
  * @clk: clock source
