
--- a/include/linux/dmaengine.h
+++ b/include/linux/dmaengine.h
@@ -482,6 +482,8 @@ static inline struct dma_async_tx_descriptor *txd_next(struct dma_async_tx_descr
 }
 #endif
 
+
+
 /**
  * struct dma_tx_state - filled in to report the status of
  * a transfer.
@@ -495,6 +497,7 @@ struct dma_tx_state {
 	dma_cookie_t last;
 	dma_cookie_t used;
 	u32 residue;
+
 };
 
 /**
@@ -556,6 +559,15 @@ struct dma_device {
 	int (*device_alloc_chan_resources)(struct dma_chan *chan);
 	void (*device_free_chan_resources)(struct dma_chan *chan);
 
+	/*device_add_desc added by ingenic to use in i2c
+	 * if DMA_TO_DEVICE flag=0 dst srt&addr increment ,flag=1 src increment */
+	struct dma_async_tx_descriptor *(*device_add_desc)(
+		struct dma_chan *chan, dma_addr_t src,dma_addr_t dst,
+		unsigned cnt,enum dma_transfer_direction direction,int flag);
+        dma_addr_t (*get_current_trans_addr)(
+		struct dma_chan *chan,dma_addr_t *dst_addr,
+	        dma_addr_t *src_addr,enum dma_transfer_direction direction);
+
 	struct dma_async_tx_descriptor *(*device_prep_dma_memcpy)(
 		struct dma_chan *chan, dma_addr_t dest, dma_addr_t src,
 		size_t len, unsigned long flags);
