
--- a/include/linux/pm_wakeup.h
+++ b/include/linux/pm_wakeup.h
@@ -99,6 +99,7 @@ extern void __pm_relax(struct wakeup_source *ws);
 extern void pm_relax(struct device *dev);
 extern void __pm_wakeup_event(struct wakeup_source *ws, unsigned int msec);
 extern void pm_wakeup_event(struct device *dev, unsigned int msec);
+extern int get_active_wakeup_sources(char *buf);
 
 #else /* !CONFIG_PM_SLEEP */
 
@@ -112,6 +113,12 @@ static inline bool device_can_wakeup(struct device *dev)
 	return dev->power.can_wakeup;
 }
 
+#ifdef CONFIG_PM
+static int get_active_wakeup_sources(char *buf)
+{
+}
+#endif
+
 static inline void wakeup_source_prepare(struct wakeup_source *ws,
 					 const char *name) {}
 
