
--- a/include/uapi/linux/soundcard.h
+++ b/include/uapi/linux/soundcard.h
@@ -566,6 +566,11 @@ typedef struct {
 #	define AFMT_MPEG		0x00000200	/* MPEG (2) audio */
 #	define AFMT_AC3		0x00000400	/* Dolby Digital AC3 */
 
+#define AFMT_S24_LE              0x00000800      /* Little endian signed 24*/
+#define AFMT_S24_BE              0x00001000      /* Big endian signed 24 */
+#define AFMT_U24_LE              0x00002000      /* Little endian unsigned 24 */
+#define AFMT_U24_BE              0x00004000      /* Big endian unsigned 24 */
+
 /*
  * Buffer status queries.
  */
