
--- a/kernel/power/main.c
+++ b/kernel/power/main.c
@@ -433,6 +433,22 @@ static ssize_t wakeup_count_store(struct kobject *kobj,
 
 power_attr(wakeup_count);
 
+static ssize_t locked_show(struct kobject *kobj,
+				struct kobj_attribute *attr,
+				char *buf)
+{
+	return get_active_wakeup_sources(buf);
+}
+
+static ssize_t locked_store(struct kobject *kobj,
+				struct kobj_attribute *attr,
+				const char *buf, size_t n)
+{
+	return 0;
+}	
+
+power_attr(locked);
+
 #ifdef CONFIG_PM_AUTOSLEEP
 static ssize_t autosleep_show(struct kobject *kobj,
 			      struct kobj_attribute *attr,
@@ -586,6 +602,7 @@ static struct attribute * g[] = {
 #ifdef CONFIG_PM_SLEEP
 	&pm_async_attr.attr,
 	&wakeup_count_attr.attr,
+	&locked_attr.attr,
 #ifdef CONFIG_PM_AUTOSLEEP
 	&autosleep_attr.attr,
 #endif
