
--- /dev/null
+++ b/sound/oss/jzsound/devices/codecs/Makefile
@@ -0,0 +1,14 @@
+# Makefile for the xburst sound oss driver
+#
+# 24 APR 2012, <jbbi: jbbi@ingernic.cn>
+#
+
+obj-$(CONFIG_JZ4780_INTERNAL_CODEC) += jz4780_codec.o jz4780_route_conf.o
+obj-$(CONFIG_HALLEY_INTERNAL_CODEC_V11) += jz_codec_v11.o jz_route_conf_v11.o
+obj-$(CONFIG_M200_INTERNAL_CODEC_V12) += jz_codec_v12.o jz_route_conf_v12.o
+obj-$(CONFIG_T10_INTERNAL_CODEC) += jz_t10_codec.o jz_route_conf_v12.o
+#obj-y += jz4775_codec.o jz4775_route_conf.o
+obj-$(CONFIG_HDMI_JZ4780)			+= hdmi_virual_codec.o
+#obj-y			+= hdmi_virual_codec.o
+obj-$(CONFIG_WM8594_CODEC_V12)			+= wm8594_codec_v12.o
+obj-$(CONFIG_AK4951_CODEC)				+= ak4951_codec.o
