
--- /dev/null
+++ b/tools/dmic-test/headers.h
@@ -0,0 +1,18 @@
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <fcntl.h>
+#include <sys/ioctl.h>
+#include <sys/soundcard.h>
+#include <stdio.h>
+#include <unistd.h>
+#include <math.h>
+#include <string.h>
+#include <stdlib.h>
+#include <sys/mman.h>
+#include <linux/fb.h>
+#include <getopt.h>
+#include <signal.h>
+#include <errno.h>
+#include <pthread.h>
+#include <stdint.h>
+
