diff -Naur uboot.ori/boards.cfg uboot/boards.cfg
--- uboot.ori/boards.cfg	2024-07-26 16:33:53.693447169 +0200
+++ uboot/boards.cfg	2024-07-26 16:48:35.815201399 +0200
@@ -62,6 +62,7 @@
 
 isvp_t31_sfcnor             mips   xburst   isvp_t31    ingenic   t31   isvp_t31:SPL_SFC_SUPPORT,ENV_IS_IN_SPI_FLASH,SPL_SFC_NOR,JZ_MMC_MSC0
 isvp_t31_sfcnor_ddr128M     mips   xburst   isvp_t31    ingenic   t31   isvp_t31:SPL_SFC_SUPPORT,ENV_IS_IN_SPI_FLASH,SPL_SFC_NOR,JZ_MMC_MSC0,DDR2_128M
+isvp_t31_sfcnor_msc1_ddr128M mips   xburst   isvp_t31    ingenic   t31   isvp_t31:SPL_SFC_SUPPORT,ENV_IS_IN_SPI_FLASH,SPL_SFC_NOR,JZ_MMC_MSC0,JZ_MMC_MSC1,DDR2_128M
 isvp_t31_sfcnor_lite        mips   xburst   isvp_t31    ingenic   t31   isvp_t31:SPL_SFC_SUPPORT,ENV_IS_IN_SPI_FLASH,SPL_SFC_NOR,JZ_MMC_MSC0,LITE_VERSION
 isvp_t31a_sfcnor_ddr128M    mips   xburst   isvp_t31    ingenic   t31   isvp_t31:SPL_SFC_SUPPORT,ENV_IS_IN_SPI_FLASH,SPL_SFC_NOR,JZ_MMC_MSC0,DDR3_128M,T31A
 isvp_t31al_sfcnor_ddr128M   mips   xburst   isvp_t31    ingenic   t31   isvp_t31:SPL_SFC_SUPPORT,ENV_IS_IN_SPI_FLASH,SPL_SFC_NOR,JZ_MMC_MSC0,DDR2_128M,T31AL
diff -Naur uboot.ori/drivers/gpio/jz_gpio/t31_gpio.c uboot/drivers/gpio/jz_gpio/t31_gpio.c
--- uboot.ori/drivers/gpio/jz_gpio/t31_gpio.c	2024-07-23 16:51:21.000000000 +0200
+++ uboot/drivers/gpio/jz_gpio/t31_gpio.c	2024-07-26 11:31:11.784136850 +0200
@@ -19,6 +19,9 @@
 #if defined(CONFIG_JZ_MMC_MSC0_PB)
 	{.port = GPIO_PORT_B, .func = GPIO_FUNC_0, .pins = (0x3 << 4 | 0xf << 0)},
 #endif
+#if defined(CONFIG_JZ_MMC_MSC1_PB)
+	{.port = GPIO_PORT_B, .func = GPIO_FUNC_1, .pins = (0x3 << 13|0xf << 8)},
+#endif
 #if defined(CONFIG_JZ_SFC_PA)
 	{.port = GPIO_PORT_A, .func = GPIO_FUNC_1, .pins = (0x3 << 23) | (0x3 << 27), .driver_strength = 4},
 #endif
diff -Naur uboot.ori/include/configs/isvp_t31.h uboot/include/configs/isvp_t31.h
--- uboot.ori/include/configs/isvp_t31.h	2024-07-23 16:51:21.000000000 +0200
+++ uboot/include/configs/isvp_t31.h	2024-07-26 11:40:16.704840420 +0200
@@ -96,10 +96,16 @@
 #define CONFIG_JZ_MMC			1
 #endif  /* JZ_MMC_MSC0 || JZ_MMC_MSC1 */
 
+#if defined(CONFIG_JZ_MMC_MSC0) && defined(CONFIG_JZ_MMC_MSC1)
+#define CONFIG_JZ_MMC_SPLMSC		1
+#define CONFIG_JZ_MMC_MSC0_PB		1
+#define CONFIG_JZ_MMC_MSC1_PB		1
+#else
 #if defined(CONFIG_JZ_MMC_MSC0)
 #define CONFIG_JZ_MMC_SPLMSC		0
 #define CONFIG_JZ_MMC_MSC0_PB		1
 #endif
+#endif  /* JZ_MMC_MSC0 && JZ_MMC_MSC1 */
 
 /*
 #if defined(CONFIG_SFC_COMMAND)
