
--- a/arch/mips/cpu/xburst/Makefile	2013-07-23 07:58:13.000000000 -0400
+++ b/arch/mips/cpu/xburst/Makefile	2023-07-26 11:31:37.000000000 -0400
@@ -26,9 +26,15 @@
 
 START	= start.o
 SOBJS-y	=
-COBJS-y	= cpu.o timer.o jz_serial.o
+COBJS-y	= cpu.o jz_serial.o timer.o
+COBJS-$(CONFIG_XBURST_TRAPS) += traps.o
+COBJS-$(CONFIG_SPL_OS_BOOT) += os_boot.o
 
-COBJS-$(CONFIG_JZ4740) += jz4740.o
+ifdef	CONFIG_SPL_BUILD
+ifdef	CONFIG_SPL_NO_CPU_SUPPORT_CODE
+START	:=
+endif
+endif
 
 SRCS	:= $(START:.o=.S) $(SOBJS-y:.o=.S) $(COBJS-y:.o=.c)
 OBJS	:= $(addprefix $(obj),$(SOBJS-y) $(COBJS-y))
