
--- a/arch/mips/include/asm/global_data.h	2013-07-23 07:58:13.000000000 -0400
+++ b/arch/mips/include/asm/global_data.h	2023-07-26 11:31:38.000000000 -0400
@@ -25,10 +25,11 @@
 #define __ASM_GBL_DATA_H
 
 #include <asm/regdef.h>
+#include <asm/global_info.h>
 
 /* Architecture-specific global data */
 struct arch_global_data {
-#ifdef CONFIG_JZSOC
+#ifdef CONFIG_JZ4740
 	/* There are other clocks in the jz4740 */
 	unsigned long per_clk;	/* Peripheral bus clock */
 	unsigned long dev_clk;	/* Device clock */
@@ -36,6 +37,9 @@
 	unsigned long tbl;
 	unsigned long lastinc;
 #endif
+#ifdef CONFIG_T23
+	struct global_info *gi;
+#endif
 };
 
 #include <asm-generic/global_data.h>
