
--- a/board/ingenic/columba/partitions.tab	1969-12-31 19:00:00.000000000 -0500
+++ b/board/ingenic/columba/partitions.tab	2023-07-26 11:31:38.000000000 -0400
@@ -0,0 +1,16 @@
+property:
+    disk_size = 7181m
+    gpt_header_lba = 512
+    custom_signature = 0
+
+partition:
+	#name     =  start,   size, fstype
+	xboot     =     0m,     3m,
+	boot      =     3m,     8m, EMPTY
+	test      =     12m,    16m, LINUX_FS
+	system    =    32m,    256m, LINUX_FS
+    data      =    288m,   32m, LINUX_FS
+    misc      =    320m,   6860m, LINUX_FS
+
+#fstype could be: LINUX_FS, FAT_FS, EMPTY
+# emcp: H9TP32A4GDCCPR
