
--- a/board/ingenic/fornax/partitions.tab	1969-12-31 19:00:00.000000000 -0500
+++ b/board/ingenic/fornax/partitions.tab	2023-07-26 11:31:38.000000000 -0400
@@ -0,0 +1,14 @@
+property:
+    disk_size = 3632m
+    gpt_header_lba = 512
+    custom_signature = 0
+
+partition:
+	#name     =  start,   size, fstype
+	xboot     =     0m,     3m,
+	boot      =     3m,     8m, EMPTY
+	system    =    12m,    64m, LINUX_FS
+        data      =    80m,  3552m, LINUX_FS
+
+#fstype could be: LINUX_FS, FAT_FS, EMPTY
+# emcp: H9TP32A4GDCCPR
\ No newline at end of file
