
--- a/board/ingenic/palladium_m200/u-boot-spl.lds	1969-12-31 19:00:00.000000000 -0500
+++ b/board/ingenic/palladium_m200/u-boot-spl.lds	2023-07-26 11:31:38.000000000 -0400
@@ -0,0 +1,52 @@
+MEMORY { .sram : ORIGIN = CONFIG_SPL_TEXT_BASE,\
+		LENGTH = CONFIG_SPL_MAX_SIZE }
+
+MEMORY { .ginfo : ORIGIN = CONFIG_SPL_GINFO_BASE,\
+		LENGTH = CONFIG_SPL_GINFO_SIZE }
+
+OUTPUT_ARCH(mips)
+ENTRY(_start)
+SECTIONS
+{
+	.text      :
+	{
+		__start = .;
+		arch/mips/cpu/xburst/m200/start.o	(.text*)
+		*(.text*)
+	} >.sram
+
+	. = ALIGN(4);
+	.rodata : { *(SORT_BY_ALIGNMENT(.rodata*)) } >.sram
+
+	. = ALIGN(4);
+	.data : { *(SORT_BY_ALIGNMENT(.data*)) } >.sram
+
+	. = ALIGN(4);
+	__image_copy_end = .;
+
+	.bss : {
+		. = ALIGN(4);
+		__bss_start = .;
+		*(.sbss.*)
+		*(.bss.*)
+		*(COMMON)
+		. = ALIGN(4);
+		__bss_end = .;
+	} >.sram
+
+	. = ALIGN(4);
+	.ginfo : { *(SORT_BY_ALIGNMENT(.ginfo*)) } >.ginfo
+
+	/DISCARD/ : {
+		*(.dynbss)
+		*(.dynstr)
+		*(.dynamic)
+		*(.interp)
+		*(.hash)
+		*(.gnu.*)
+		*(.plt)
+		*(.got.plt)
+		*(.rel.plt)
+		*(.rel.dyn)
+	}
+}
