
--- a/common/cmd_sf.c	2013-07-23 07:58:13.000000000 -0400
+++ b/common/cmd_sf.c	2024-01-20 19:46:04.453045675 -0500
@@ -13,20 +13,24 @@
 #include <asm/io.h>
 
 #ifndef CONFIG_SF_DEFAULT_SPEED
-# define CONFIG_SF_DEFAULT_SPEED	1000000
+#define CONFIG_SF_DEFAULT_SPEED	1000000
 #endif
+
 #ifndef CONFIG_SF_DEFAULT_MODE
-# define CONFIG_SF_DEFAULT_MODE		SPI_MODE_3
+#define CONFIG_SF_DEFAULT_MODE		SPI_MODE_3
 #endif
+
 #ifndef CONFIG_SF_DEFAULT_CS
-# define CONFIG_SF_DEFAULT_CS		0
+#define CONFIG_SF_DEFAULT_CS		0
 #endif
+
 #ifndef CONFIG_SF_DEFAULT_BUS
-# define CONFIG_SF_DEFAULT_BUS		0
+#define CONFIG_SF_DEFAULT_BUS		0
 #endif
 
 static struct spi_flash *flash;
 
+#define PRINT_TIME
 
 /*
  * This function computes the length argument for the erase command.
@@ -483,6 +487,11 @@
 	--argc;
 	++argv;
 
+#ifdef PRINT_TIME
+	unsigned int start, end;
+	start = get_timer(0);
+#endif
+
 	if (strcmp(cmd, "probe") == 0) {
 		ret = do_spi_flash_probe(argc, argv);
 		goto done;
@@ -507,9 +516,13 @@
 		ret = -1;
 
 done:
+#ifdef PRINT_TIME
+	end = get_timer(0);
+	printf("--->%s spend %d ms\n",cmd,end - start);
+#endif
+
 	if (ret != -1)
 		return ret;
-
 usage:
 	return CMD_RET_USAGE;
 }
