
--- a/common/fdt_support.c	2013-07-23 07:58:13.000000000 -0400
+++ b/common/fdt_support.c	2023-07-26 11:31:38.000000000 -0400
@@ -498,6 +498,7 @@
 	uint64_t addr, size;
 	int total, ret;
 	uint actualsize;
+	int fdt_memrsv = 0;
 
 	if (!blob)
 		return 0;
@@ -507,6 +508,7 @@
 		fdt_get_mem_rsv(blob, i, &addr, &size);
 		if (addr == (uintptr_t)blob) {
 			fdt_del_mem_rsv(blob, i);
+			fdt_memrsv = 1;
 			break;
 		}
 	}
@@ -528,9 +530,11 @@
 	fdt_set_totalsize(blob, actualsize);
 
 	/* Add the new reservation */
-	ret = fdt_add_mem_rsv(blob, (uintptr_t)blob, actualsize);
-	if (ret < 0)
-		return ret;
+	if (fdt_memrsv) {
+		ret = fdt_add_mem_rsv(blob, (uintptr_t)blob, actualsize);
+		if (ret < 0)
+			return ret;
+	}
 
 	return actualsize;
 }
