
--- a/common/spl/spl.c	2013-07-23 07:58:13.000000000 -0400
+++ b/common/spl/spl.c	2024-01-20 19:34:09.672623577 -0500
@@ -32,6 +32,7 @@
 #include <image.h>
 #include <malloc.h>
 #include <linux/compiler.h>
+#include <regulator.h>
 
 DECLARE_GLOBAL_DATA_PTR;
 
@@ -123,7 +124,7 @@
 	image_entry_noargs_t image_entry =
 			(image_entry_noargs_t) spl_image->entry_point;
 
-	debug("image entry point: 0x%X\n", spl_image->entry_point);
+	printf("image entry point: 0x%X\n", spl_image->entry_point);
 	image_entry();
 }
 
@@ -165,9 +166,12 @@
 #ifdef CONFIG_SPL_BOARD_INIT
 	spl_board_init();
 #endif
-
 	boot_device = spl_boot_device();
 	debug("boot device - %d\n", boot_device);
+#ifdef CONFIG_PALLADIUM
+	spl_board_prepare_for_linux();
+#endif
+
 	switch (boot_device) {
 #ifdef CONFIG_SPL_RAM_DEVICE
 	case BOOT_DEVICE_RAM:
@@ -181,7 +185,19 @@
 		spl_mmc_load_image();
 		break;
 #endif
-#ifdef CONFIG_SPL_NAND_SUPPORT
+#ifdef CONFIG_SPL_SFC_SUPPORT
+#ifdef CONFIG_SFC_NOR
+	case BOOT_DEVICE_SFC_NOR:
+		spl_sfc_nor_load_image();
+		break;
+#endif
+#ifdef CONFIG_SFC_NAND
+	case BOOT_DEVICE_SFC_NAND:
+		spl_sfc_nand_load_image();
+		break;
+#endif
+#endif
+#if defined(CONFIG_SPL_NAND_SUPPORT) || defined(CONFIG_JZ_NAND_MGR)
 	case BOOT_DEVICE_NAND:
 		spl_nand_load_image();
 		break;
@@ -238,6 +254,7 @@
 	default:
 		debug("Unsupported OS image.. Jumping nevertheless..\n");
 	}
+
 	jump_to_image_no_args(&spl_image);
 }
 
@@ -247,16 +264,45 @@
  */
 void preloader_console_init(void)
 {
+	unsigned int chip = 0;
 	gd->bd = &bdata;
+#ifndef CONFIG_BURNER
 	gd->baudrate = CONFIG_BAUDRATE;
-
+#else
+	gd->baudrate = gd->arch.gi->baud_rate;
+#endif
+#ifdef CONFIG_PALLADIUM
+	gd->baudrate = 3750000;
+#endif
 	serial_init();		/* serial communications setup */
 
 	gd->have_console = 1;
 
-	puts("\nU-Boot SPL " PLAIN_VERSION " (" U_BOOT_DATE " - " \
+	puts("\n\nU-Boot SPL " PLAIN_VERSION " (" U_BOOT_DATE " - " \
 			U_BOOT_TIME ")\n");
+	chip = *((volatile unsigned int *)(0xb3540238));
+	chip = chip >> 16;
+	if (0x1111 == chip) {
+		puts("Board info: T23N\n");
+	} else if (0x6666 == chip) {
+		puts("Board info: T23ZN\n");
+	} else {
+		puts("Board info: No SOC Info\n");
+	}
 #ifdef CONFIG_SPL_DISPLAY_PRINT
 	spl_display_print();
 #endif
 }
+
+void spl_regulator_set(void)
+{
+#ifdef CONFIG_SPL_CORE_VOLTAGE
+	spl_regulator_init();
+	debug("Set core voltage:%dmv\n", CONFIG_SPL_CORE_VOLTAGE);
+	spl_regulator_set_voltage(REGULATOR_CORE, CONFIG_SPL_CORE_VOLTAGE);
+#endif
+#ifdef CONFIG_SPL_MEM_VOLTAGE
+	debug("Set mem voltage:%dmv\n", CONFIG_SPL_MEM_VOLTAGE);
+	spl_regulator_set_voltage(REGULATOR_MEM, CONFIG_SPL_MEM_VOLTAGE);
+#endif
+}
