
--- a/drivers/mtd/spi/spi_flash_internal.h	2013-07-23 07:58:13.000000000 -0400
+++ b/drivers/mtd/spi/spi_flash_internal.h	2023-07-26 11:31:38.000000000 -0400
@@ -35,6 +35,11 @@
 #define SPI_FLASH_SPANSION_IDCODE0	0x01
 #define SPI_FLASH_STMICRO_IDCODE0	0x20
 #define SPI_FLASH_WINBOND_IDCODE0	0xef
+#define SPI_FLASH_GD25_IDCODE0      0xc8
+#define SPI_FLASH_MX25_IDCODE0      0xc2
+#define SPI_FLASH_EN25_IDCODE0      0x1c
+#define SPI_FLASH_XT25_IDCODE0		0x0b
+#define SPI_FLASH_XM25_IDCODE0		0x20
 
 #ifdef CONFIG_SPI_FLASH_BAR
 /* Bank addr access commands */
@@ -129,6 +134,8 @@
 int spi_flash_cmd_erase(struct spi_flash *flash, u32 offset, size_t len);
 
 /* Manufacturer-specific probe functions */
+struct spi_flash *spi_flash_probe_ingenic(struct spi_slave *spi, u8 *idcode);
+struct spi_flash *spi_flash_probe_ingenic_nand(struct spi_slave *spi, u8 *idcode);
 struct spi_flash *spi_flash_probe_spansion(struct spi_slave *spi, u8 *idcode);
 struct spi_flash *spi_flash_probe_atmel(struct spi_slave *spi, u8 *idcode);
 struct spi_flash *spi_flash_probe_eon(struct spi_slave *spi, u8 *idcode);
