
--- a/drivers/nand/driver/inc/nderrno.h	1969-12-31 19:00:00.000000000 -0500
+++ b/drivers/nand/driver/inc/nderrno.h	2023-07-26 11:31:38.000000000 -0400
@@ -0,0 +1,16 @@
+#ifndef _NAND_ERRNO_H_
+#define _NAND_ERRNO_H_
+
+#include "pagelist.h"
+
+#define BLOCK_MOVE  	ND_ECC_TOOLARGE
+#define SUCCESS         0
+#define ENAND		ND_ERROR_ARGS
+#define DMA_AR  	ND_ERROR_MEMORY
+#define IO_ERROR      	ND_ERROR_IO
+#define TIMEOUT       	ND_TIMEOUT
+#define ECC_ERROR     	ND_ERROR_ECC
+#define ALL_FF	 	ND_ERROR_NOWRITE
+#define WRITE_PROTECT	-7
+
+#endif /*_NAND_ERRNO_H_*/
