
--- a/drivers/nand/manager/inc/os/NandAlloc.h	1969-12-31 19:00:00.000000000 -0500
+++ b/drivers/nand/manager/inc/os/NandAlloc.h	2023-07-26 11:31:38.000000000 -0400
@@ -0,0 +1,10 @@
+#ifndef _NANDALLOC_H_
+#define _NANDALLOC_H_
+void* Nand_MemoryInit(void *h,int size,int type);
+void Nand_MemoryDeinit(void);
+void *Nand_VirtualAlloc(int size);
+void Nand_VirtualFree(void *val);
+void *Nand_ContinueAlloc(int size);
+void Nand_ContinueFree(void *val);
+
+#endif /* _NANDALLOC_H_ */
