
--- a/drivers/nand/manager/inc/os/NandDebug.h	1969-12-31 19:00:00.000000000 -0500
+++ b/drivers/nand/manager/inc/os/NandDebug.h	2023-07-26 11:31:38.000000000 -0400
@@ -0,0 +1,11 @@
+#ifndef _NANDDEBUG_H_
+#define _NANDDEBUG_H_
+
+extern int nm_dbg_level;
+extern int utils_dbg_level;
+extern int libops_dbg_level;
+
+void nd_dump_stack(void);
+int __ndprint(const char *s, ...) __attribute__ ((format(printf, 1, 2)));
+
+#endif /* _NANDDEBUG_H_ */
