
--- a/drivers/nand/manager/inc/sectorlist.h	1969-12-31 19:00:00.000000000 -0500
+++ b/drivers/nand/manager/inc/sectorlist.h	2023-07-26 11:31:38.000000000 -0400
@@ -0,0 +1,21 @@
+#ifndef __SECTORLIST_H__
+#define __SECTORLIST_H__
+
+#include "singlelist.h"
+
+typedef struct _SectorList SectorList;
+struct _SectorList {
+	struct singlelist head;//must be the first member of the struct
+	unsigned int startSector;
+	int sectorCount;
+	void *pData;
+};
+
+#define copysectorlist(src,tar)						\
+	do{												\
+		src->startSector = tar->startSector;		\
+		src->sectorCount = tar->sectorCount;		\
+		src->pData = tar->pData;					\
+	}while(0)
+
+#endif
