
--- a/drivers/spi/Makefile	2013-07-23 07:58:13.000000000 -0400
+++ b/drivers/spi/Makefile	2023-07-26 11:31:38.000000000 -0400
@@ -48,6 +48,9 @@
 COBJS-$(CONFIG_OC_TINY_SPI) += oc_tiny_spi.o
 COBJS-$(CONFIG_OMAP3_SPI) += omap3_spi.o
 COBJS-$(CONFIG_SOFT_SPI) += soft_spi.o
+COBJS-$(CONFIG_INGENIC_SOFT_SPI) += ingenic_soft_spi.o
+COBJS-$(CONFIG_JZ_SPI) += jz_spi.o
+COBJS-$(CONFIG_SFC_NOR) += jz_sfc_nor.o
 COBJS-$(CONFIG_SH_SPI) += sh_spi.o
 COBJS-$(CONFIG_FSL_ESPI) += fsl_espi.o
 COBJS-$(CONFIG_FDT_SPI) += fdt_spi.o
