
--- a/include/ddr/ddr_common.h	1969-12-31 19:00:00.000000000 -0500
+++ b/include/ddr/ddr_common.h	2023-07-26 11:31:38.000000000 -0400
@@ -0,0 +1,13 @@
+#ifndef __DDR_COMMON_H__
+#define __DDR_COMMON_H__
+
+#include <ddr/ddr_chips.h>
+#include <ddr/ddr_params.h>
+
+#ifdef CONFIG_CPU_XBURST
+#include <asm/ddr_dwc.h>
+#include <ddr/ddrc.h>
+#include <ddr/ddrp_dwc.h>
+#endif
+
+#endif /* __DDR_COMMON_H__ */
