
--- a/include/ddr-fpga/ddr_chips.h	1969-12-31 19:00:00.000000000 -0500
+++ b/include/ddr-fpga/ddr_chips.h	2023-07-26 11:31:38.000000000 -0400
@@ -0,0 +1,59 @@
+#ifndef __DDR_CHIPS_H__
+#define __DDR_CHIPS_H__
+
+#ifdef CONFIG_MDDR_W94AD6KB_HX51
+#include "./chips/MDDR_W94AD6KB-HX51.h"
+#endif
+
+#ifdef CONFIG_DDR3_H5TQ1G83DFR_H9C
+#include "./chips/DDR3_H5TQ1G83DFR_H9C.h"
+#endif
+
+#ifdef CONFIG_DDR3_H5TQ2G83CFR_H9C
+#include "./chips/DDR3_H5TQ2G83CFR_H9C.h"
+#endif
+
+#ifdef CONFIG_DDR3_H5TQ2G63BFR_H9C
+#include "./chips/DDR3_H5TQ2G63BFR_H9C.h"
+#endif
+
+#ifdef CONFIG_DDR3_TSD34096M1333C9_E
+#include "./chips/DDR3_TSD34096M1333C9_E.h"
+#endif
+
+#ifdef CONFIG_MCP_KMS5U000JM_B308
+#include "./chips/MCP_KMS5U000JM_B308.h"
+#endif
+
+#ifdef CONFIG_MCP_KMS5U000KM-B308
+#include "./chips/MCP_KMS5U000KM-B308.h"
+#endif
+
+#ifdef CONFIG_MCP_H9TP32A8JDMC_PRKGM_LPDDR2
+#include "./chips/MCP_H9TP32A8JDMC_PRKGM_LPDDR2.h"
+#endif
+
+#ifdef CONFIG_FMT4D32UAB_25LC_LPDDR2
+#include "./chips/FMT4D32UAB_25LC_LPDDR2.h"
+#endif
+
+#ifdef CONFIG_MCP_SAMSUNG_KMN5X000ZM_LPDDR2
+#include "./chips/MCP_SAMSUNG_KMN5X000ZM_LPDDR2.h"
+#endif
+
+#ifdef CONFIG_DDR2_PWE809416BBR_E7DN
+#include "./chips/DDR2_PWE809416BBR-E7DN.h"
+#endif
+
+#ifdef CONFIG_MDDR_H5MS5122DFR_J3M
+#include "./chips/MDDR_H5MS5122DFR_J3M.h"
+#endif
+
+#ifdef CONFIG_DDR2_M14D5121632A
+#include "./chips/DDR2_M14D5121632A.h"
+#endif
+
+#ifdef CONFIG_DDR2_M14D1G1664A.h
+#include "./chips/DDR2_M14D1G1664A.h"
+#endif
+#endif /* __DDR_CHIPS_H__ */
