
--- a/include/ingenic_soft_spi.h	1969-12-31 19:00:00.000000000 -0500
+++ b/include/ingenic_soft_spi.h	2023-07-26 11:31:38.000000000 -0400
@@ -0,0 +1,15 @@
+
+#ifndef __INCLUDE__GPIO__SPI__
+#define __INCLUDE__GPIO__SPI__
+
+struct spi {
+	unsigned int clk ;
+	unsigned int data_in ;
+	unsigned int data_out ;
+	unsigned int enable  ;
+	unsigned int rate  ;
+};
+
+void spi_init_jz(struct spi *spi);
+
+#endif
