
--- a/include/rle_logo.h	1969-12-31 19:00:00.000000000 -0500
+++ b/include/rle_logo.h	2023-07-26 11:31:38.000000000 -0400
@@ -0,0 +1,12 @@
+#ifndef _RLE_LOGO_H
+#define _RLE_LOGO_H
+#include <rle_boot_logo.h>
+#define RLE_LOGO_DEFAULT_ADDR  rle_default_logo_addr	//need to fixed!
+#if !defined(CONFIG_LCD_INFO_BELOW_LOGO)
+#define  BMP_LOGO_HEIGHT  panel_info.vl_row
+#define  BMP_LOGO_WIDTH   panel_info.vl_col
+#else
+#define  BMP_LOGO_HEIGHT  0
+#define  BMP_LOGO_WIDTH   0
+#endif
+#endif
