
--- a/LOG/cmi_mpc5xx.ERR	2023-01-02 16:15:55.884137657 -0500
+++ b/LOG/cmi_mpc5xx.ERR	1969-12-31 19:00:00.000000000 -0500
@@ -1,20 +0,0 @@
-mips-linux-gnu-gcc: error: unrecognized command line option '-mrelocatable'
-mips-linux-gnu-gcc: error: unrecognized command line option '-meabi'
-mips-linux-gnu-gcc: error: unrecognized command line option '-mpowerpc'
-mips-linux-gnu-gcc: error: unrecognized command line option '-mrelocatable'
-mips-linux-gnu-gcc: error: unrecognized command line option '-meabi'
-mips-linux-gnu-gcc: error: unrecognized command line option '-mpowerpc'
-mips-linux-gnu-gcc: error: unrecognized command line option '-mrelocatable'
-mips-linux-gnu-gcc: error: unrecognized command line option '-meabi'
-mips-linux-gnu-gcc: error: unrecognized command line option '-mpowerpc'
-mips-linux-gnu-gcc: error: unrecognized command line option '-mrelocatable'
-mips-linux-gnu-gcc: error: unrecognized command line option '-meabi'
-mips-linux-gnu-gcc: error: unrecognized command line option '-mpowerpc'
-mips-linux-gnu-gcc: error: unrecognized command line option '-mrelocatable'
-mips-linux-gnu-gcc: error: unrecognized command line option '-meabi'
-mips-linux-gnu-gcc: error: unrecognized command line option '-mpowerpc'
-mips-linux-gnu-gcc: error: unrecognized command line option '-mrelocatable'
-mips-linux-gnu-gcc: error: unrecognized command line option '-meabi'
-mips-linux-gnu-gcc: error: unrecognized command line option '-mpowerpc'
-make: *** [Makefile:715: lib/asm-offsets.s] Error 1
-make: *** Waiting for unfinished jobs....
