
--- a/tools/ingenic-tools/nand_device/sky_nand.c	1969-12-31 19:00:00.000000000 -0500
+++ b/tools/ingenic-tools/nand_device/sky_nand.c	2023-07-26 11:31:38.000000000 -0400
@@ -0,0 +1,23 @@
+#include <stdio.h>
+#include "nand_common.h"
+
+#define SKY_MID			    0x01
+#define SKY_NAND_DEVICD_COUNT	    1
+
+static unsigned char sky_errstat[]= {0x3};
+
+static struct device_struct device[] = {
+	DEVICE_STRUCT(0x15, 2048, 2, 4, 2, 1, sky_errstat),
+};
+
+static struct nand_desc sky_nand = {
+
+	.id_manufactory = SKY_MID,
+	.device_counts = SKY_NAND_DEVICD_COUNT,
+	.device = device,
+};
+
+int sky_nand_register_func(void) {
+	return nand_register(&sky_nand);
+}
+
