
--- a/tools/ingenic-tools-fpag/nand_device/foresee_nand.c	1969-12-31 19:00:00.000000000 -0500
+++ b/tools/ingenic-tools-fpag/nand_device/foresee_nand.c	2023-07-26 11:31:38.000000000 -0400
@@ -0,0 +1,27 @@
+#include <stdio.h>
+#include "nand_common.h"
+
+#define ATO_MID			    0xCD
+#define ATO_NAND_DEVICD_COUNT	    3
+
+static unsigned char fw_xaw[] = {0x02};
+static unsigned char fw_xbw[] = {0x02};
+static unsigned char fs_xaw[] = {0x01, 0x02};
+
+
+static struct device_struct device[] = {
+	DEVICE_STRUCT(0xB1, 2048, 2, 4, 3, 1, fw_xaw),
+	DEVICE_STRUCT(0xEB, 2048, 2, 4, 2, 1, fw_xaw),
+	DEVICE_STRUCT(0x71, 2048, 2, 4, 2, 2, fs_xaw),
+};
+
+static struct nand_desc foresee_nand = {
+
+	.id_manufactory = ATO_MID,
+	.device_counts = ATO_NAND_DEVICD_COUNT,
+	.device = device,
+};
+
+int foresee_nand_register_func(void) {
+	return nand_register(&foresee_nand);
+}
