
--- a/tools/ingenic-tools-fpag/nand_device/xtx_mid2c_nand.c	1969-12-31 19:00:00.000000000 -0500
+++ b/tools/ingenic-tools-fpag/nand_device/xtx_mid2c_nand.c	2023-07-26 11:31:38.000000000 -0400
@@ -0,0 +1,22 @@
+#include <stdio.h>
+#include "nand_common.h"
+
+#define XTX_MID2C_MID			    0x2C
+#define XTX_MID2C_NAND_DEVICD_COUNT	    1
+
+static unsigned char xtx_mid2c_xaw[] = {0x2};
+
+static struct device_struct device[1] = {
+	DEVICE_STRUCT(0x24, 2048, 2, 4, 3, 1, xtx_mid2c_xaw),
+};
+
+static struct nand_desc xtx_mid2c_nand = {
+
+	.id_manufactory = XTX_MID2C_MID,
+	.device_counts = XTX_MID2C_NAND_DEVICD_COUNT,
+	.device = device,
+};
+
+int xtx_mid2c_nand_register_func(void) {
+	return nand_register(&xtx_mid2c_nand);
+}
